[QoL][Feature] Hall of Fame + Run History

frutescens opened the pull request:
## What are the changes?
This adds a new option on the menu for 'Run History'.
It allows users to store and look at a maximum of 25 past runs. This functionality automatically logs a run based on the current SessionSaveData and stores it in cache/the network database. At the moment, the function mirrors the 'Load Save' UI and only shows the team, whether they were victorious or not, the mode, wave, and the final timestamp of the run.

The server-side code is here: https://github.com/pagefaultgames/rogueserver/pull/33
The previous PR (closed because of beta collision): https://github.com/pagefaultgames/pokerogue/pull/1815

## Why am I doing these changes?
This has been a feature requested by many users. In addition, this will allow the Balance and Development teams to do some analytics and have data on the gameplay experience.

I also have a bad memory so I thought it would be helpful to be able to access past runs to see what I need to do better.

## What did change?
- game-data.ts : 
- menu-ui-handler.ts :
- phases.ts :
- run-history-ui-handler.ts :
- run-info-ui-handler.ts :
- loading-scene.ts : 
- New localization files added for run-history specifically

### Screenshots/Videos (08/16/24)
English
![Uploading image.png…]()


German

French
![image](https://github.com/user-attachments/assets/fb32be00-9906-4eb7-bc24-f441f769d93b)
![image](https://github.com/user-attachments/assets/e5d33dec-697c-4a55-a604-b802870d20ab)
![image](https://github.com/user-attachments/assets/1ad32ad1-e722-4abf-a3fc-2e4986c23bf8)

Italian

Spanish
![image](https://github.com/user-attachments/assets/5df62060-78f5-4e87-a01e-4ee2264330f2)
![Uploading image.png…]()


Portuguese 

Korean

Simplified Chinese
![image](https://github.com/user-attachments/assets/f688fb4a-0f37-4d1a-97ec-9b0b709046c0)
![image](https://github.com/user-attachments/assets/4385683a-e922-4e78-b573-d7a30bdbee2f)
![image](https://github.com/user-attachments/assets/5bfe7b04-83dd-4f86-88db-6480a7605d9b)

Japanese

## How to test the changes?
Play the game normally and the game should record defeats / victories. 

##Localization Checklist
Thank you to everyone who contributed! Much appreciated. 
- [x] Spanish
- [x] German
- [x] French
- [x] Brazilian Portuguese 
- [x] Italian
- [x] Korean
- [x] Traditional Chinese
- [x] Simplified Chinese
- [x] Japanese 

## Checklist
- [x] **I'm using `beta` as my base branch**
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I considered writing automated tests for the issue?
- [x] Have I tested the changes (manually)?
    - [x] Are all unit tests still passing? (`npm run test`)
- [x] Are the changes visual?
  - [x] Have I provided screenshots/videos of the changes?

CodeTappert commented:
Please make sure that all languages have the same amount of localization keys. And fill all other lanugages (that you are not a native of) with english placeholders. Thank you!

CodeTappert commented on src/locales/en/menu-ui-handler.ts:
```diff
@@ -4,6 +4,7 @@ export const menuUiHandler: SimpleTranslationEntries = {
   "GAME_SETTINGS": "Game Settings",
   "ACHIEVEMENTS": "Achievements",
   "STATS": "Stats",
+  "RUN_HISTORY": "Run History",
```
This is still missing in all other languages no?

EnochG1 commented on src/locales/ko/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,35 @@
+import { SimpleTranslationEntries } from "#app/interfaces/locales";
+
+export const runHistory: SimpleTranslationEntries = {
+  "victory": "Victory!",
+  "defeatedWild": "Defeated by ",
+  "defeatedTrainer": "Defeated by ",
+  "defeatedTrainerDouble": "Defeated by Duo",
+  "defeatedRival": "Defeated by Rival",
+  "defeated":"Defeated",
+  "luck":"Luck",
+  "score":"Score",
+  "mode":"Mode",
+  "challengeRules":"Rule(s)",
+  "challengeMonoGen1":"Gen I",
+  "challengeMonoGen2":"Gen II",
+  "challengeMonoGen3":"Gen III",
+  "challengeMonoGen4":"Gen IV",
+  "challengeMonoGen5":"Gen V",
+  "challengeMonoGen6":"Gen VI",
+  "challengeMonoGen7":"Gen VII",
+  "challengeMonoGen8":"Gen VIII",
+  "challengeMonoGen9":"Gen IX",
+  "playerItems":"Player Items",
+  "personalBest":"Personal Best!",
+  "SPDshortened":"Vel.",
+  "runInfo":"Run Info",
+  "money":"Money",
+  "runLength":"Run Length",
+  "viewHeldItems":"Held Items",
+  "hallofFameText":"Welcome to the Hall of Fame!",
+  "viewHallOfFame":"View Hall of Fame!"
```
```suggestion
  "victory": "승리!",
  "defeatedWild": "야생에서 패배: ",
  "defeatedTrainer": "트레이너에게 패배: ",
  "defeatedTrainerDouble": "더블 배틀에서 패배",
  "defeatedRival": "라이벌에게 패배",
  "defeated":"패배",
  "luck":"행운",
  "score":"점수",
  "mode":"모드",
  "challengeRules":"규칙",
  "challengeMonoGen1":"1세대",
  "challengeMonoGen2":"2세대",
  "challengeMonoGen3":"3세대",
  "challengeMonoGen4":"4세대",
  "challengeMonoGen5":"5세대",
  "challengeMonoGen6":"6세대",
  "challengeMonoGen7":"7세대",
  "challengeMonoGen8":"8세대",
  "challengeMonoGen9":"9세대",
  "playerItems":"플레이어 아이템",
  "personalBest":"개인 최고기록!",
  "SPDshortened":"스피드",
  "runInfo":"플레이 정보",
  "money":"소지금",
  "runLength":"플레이 타임",
  "viewHeldItems":"도구",
  "hallofFameText":"전당 등록을 축하합니다!",
  "viewHallOfFame":"전당 보기"
```

EnochG1 reviewed (commented):
Korean translation

EnochG1 commented on src/ui/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,334 @@
+          //otherwise it becomes Rival_5 in Ivy's case
+          this.add(gameOutcomeLabel);
+        } else if (data.trainer.variant === TrainerVariant.DOUBLE) {
+          const gameOutcomeLabel = addTextObject(this.scene, 8, 5, `${i18next.t("runHistory:defeatedTrainer")+tObj.config.nameDouble+" "+tObj.getName(0, false)}`, TextStyle.WINDOW);
```
I think this line have to get `i18next.t("runHistory:defeatedTrainerDouble")` instead of `i18next.t("runHistory:defeatedTrainer")`.
But not sure if position of tObj.config.nameDouble is wrong or not.

frutescens replied:
You're right. I will fix that.

CodeTappert commented on src/locales/de/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,35 @@
+import { SimpleTranslationEntries } from "#app/interfaces/locales";
+
+export const runHistory: SimpleTranslationEntries = {
+  "victory": "Victory!",
+  "defeatedWild": "Defeated by ",
+  "defeatedTrainer": "Defeated by ",
+  "defeatedTrainerDouble": "Defeated by Duo",
+  "defeatedRival": "Defeated by Rival",
+  "defeated":"Defeated",
+  "luck":"Luck",
+  "score":"Score",
+  "mode":"Mode",
+  "challengeRules":"Rule(s)",
+  "challengeMonoGen1":"Gen I",
+  "challengeMonoGen2":"Gen II",
+  "challengeMonoGen3":"Gen III",
+  "challengeMonoGen4":"Gen IV",
+  "challengeMonoGen5":"Gen V",
+  "challengeMonoGen6":"Gen VI",
+  "challengeMonoGen7":"Gen VII",
+  "challengeMonoGen8":"Gen VIII",
+  "challengeMonoGen9":"Gen IX",
+  "playerItems":"Player Items",
+  "personalBest":"Personal Best!",
+  "SPDshortened":"Vel.",
+  "runInfo":"Run Info",
+  "money":"Money",
+  "runLength":"Run Length",
+  "viewHeldItems":"Held Items",
+  "hallofFameText":"Welcome to the Hall of Fame!",
+  "viewHallOfFame":"View Hall of Fame!"
+} as const;
```
```suggestion
  "victory": "Sieg!",
  "defeatedWild": "Besiegt durch ",
  "defeatedTrainer": "Besiegt durch",
  "defeatedTrainerDouble": "Besiegt durch Doppelkampf",
  "defeatedRival": "Besiegt durch Rivale",
  "defeated":"Besiegt",
  "luck":"Glück",
  "score":"Punkte",
  "mode":"Modus",
  "challengeRules":"Regeln",
  "challengeMonoGen1":"Gen I",
  "challengeMonoGen2":"Gen II",
  "challengeMonoGen3":"Gen III",
  "challengeMonoGen4":"Gen IV",
  "challengeMonoGen5":"Gen V",
  "challengeMonoGen6":"Gen VI",
  "challengeMonoGen7":"Gen VII",
  "challengeMonoGen8":"Gen VIII",
  "challengeMonoGen9":"Gen IX",
  "playerItems":"Spielergegenstände",
  "personalBest":"Persönlicher Bestwert!",
  "SPDshortened":"Geschw.",
  "runInfo":"Run Info.",
  "money":"Geld",
  "runLength":"Run Länge,
  "viewHeldItems":"Getragene Items",
  "hallofFameText":"Willkommen in der Ruhmeshalle",
  "viewHallOfFame":"Ruhmeshalle ansehen!"
} as const;
```

CodeTappert commented on src/locales/de/menu-ui-handler.ts:
```diff
@@ -4,6 +4,7 @@ export const menuUiHandler: SimpleTranslationEntries = {
   "GAME_SETTINGS": "Spieleinstellungen",
   "ACHIEVEMENTS": "Erfolge",
   "STATS": "Statistiken",
+  "RUN_HISTORY": "Run History",
```
```suggestion
  "RUN_HISTORY": "Laufhistorie",
```

CodeTappert commented:
Please ping me when you merge the german changes in. So i can test if they fit etc

CodeTappert commented:
![image](https://github.com/user-attachments/assets/740a1d4e-4e09-4d64-80c4-1195373c81c2)
EXCUSE ME. ITS PERFECTION. <3

![image](https://github.com/user-attachments/assets/1dc656ad-5810-4413-9485-20d8f8493c48)

josericardo-fo commented on src/locales/pt_BR/menu-ui-handler.ts:
```diff
@@ -4,6 +4,7 @@ export const menuUiHandler: SimpleTranslationEntries = {
   "GAME_SETTINGS": "Configurações",
   "ACHIEVEMENTS": "Conquistas",
   "STATS": "Estatísticas",
+  "RUN_HISTORY": "Run History",
```
```suggestion
  "RUN_HISTORY": "Histórico de Jogos",
```

josericardo-fo commented on src/locales/pt_BR/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,35 @@
+import { SimpleTranslationEntries } from "#app/interfaces/locales";
+
+export const runHistory: SimpleTranslationEntries = {
+  "victory": "Victory!",
+  "defeatedWild": "Defeated by ",
+  "defeatedTrainer": "Defeated by ",
+  "defeatedTrainerDouble": "Defeated by Duo",
+  "defeatedRival": "Defeated by Rival",
+  "defeated":"Defeated",
+  "luck":"Luck",
+  "score":"Score",
+  "mode":"Mode",
+  "challengeRules":"Rule(s)",
+  "challengeMonoGen1":"Gen I",
+  "challengeMonoGen2":"Gen II",
+  "challengeMonoGen3":"Gen III",
+  "challengeMonoGen4":"Gen IV",
+  "challengeMonoGen5":"Gen V",
+  "challengeMonoGen6":"Gen VI",
+  "challengeMonoGen7":"Gen VII",
+  "challengeMonoGen8":"Gen VIII",
+  "challengeMonoGen9":"Gen IX",
+  "playerItems":"Player Items",
+  "personalBest":"Personal Best!",
+  "SPDshortened":"Vel.",
+  "runInfo":"Run Info",
+  "money":"Money",
+  "runLength":"Run Length",
+  "viewHeldItems":"Held Items",
+  "hallofFameText":"Welcome to the Hall of Fame!",
+  "viewHallOfFame":"View Hall of Fame!"
```
```suggestion
  "victory": "Vitória!",
  "defeatedWild": "Derrotado por ",
  "defeatedTrainer": "Derrotado por ",
  "defeatedTrainerDouble": "Derrotado por Dupla",
  "defeatedRival": "Derrotado por Rival",
  "defeated": "Derrotado",
  "luck": "Sorte",
  "score": "Pontuação",
  "mode": "Modo",
  "challengeRules": "Regra(s)",
  "challengeMonoGen1": "G1",
  "challengeMonoGen2": "G2",
  "challengeMonoGen3": "G3",
  "challengeMonoGen4": "G4",
  "challengeMonoGen5": "G5",
  "challengeMonoGen6": "G6",
  "challengeMonoGen7": "G7",
  "challengeMonoGen8": "G8",
  "challengeMonoGen9": "G9",
  "playerItems": "Itens do Jogador",
  "personalBest": "Recorde Pessoal!",
  "SPDshortened": "Vel.",
  "runInfo": "Info. do Jogo",
  "money": "Dinheiro",
  "runLength": "Tamanho do Jogo",
  "viewHeldItems": "Itens Segurados",
  "hallofFameText": "Bem-vindo(a) ao Hall da Fama!",
  "viewHallOfFame": "Veja o Hall da Fama!"
```

Adri1 reviewed (commented):
Added French
However, would it be possible to have in-game screenshots about all of this new feature? (In English, so it could but used by everyone) Because I don't want to provide a blind translation that might appear to be inaccurate because I do not know the overall environment of what I have to translate

EnochG1 commented on src/locales/ko/menu-ui-handler.ts:
```diff
@@ -4,6 +4,7 @@ export const menuUiHandler: SimpleTranslationEntries = {
   "GAME_SETTINGS": "게임 설정",
   "ACHIEVEMENTS": "업적",
   "STATS": "통계",
+  "RUN_HISTORY": "Run History",
```
```suggestion
  "RUN_HISTORY": "플레이 이력",
```

EnochG1 reviewed (commented):
Sorry, I missed translation.

40chyan commented on src/locales/zh_CN/menu-ui-handler.ts:
```diff
@@ -4,6 +4,7 @@ export const menuUiHandler: SimpleTranslationEntries = {
   "GAME_SETTINGS": "游戏设置",
   "ACHIEVEMENTS": "成就",
   "STATS": "数据统计",
+  "RUN_HISTORY": "Run History",
```
```suggestion
  "RUN_HISTORY": "历史记录",
```

40chyan commented on src/locales/zh_CN/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,35 @@
+import { SimpleTranslationEntries } from "#app/interfaces/locales";
+
+export const runHistory: SimpleTranslationEntries = {
+  "victory": "Victory!",
+  "defeatedWild": "Defeated by ",
+  "defeatedTrainer": "Defeated by ",
+  "defeatedTrainerDouble": "Defeated by Duo",
+  "defeatedRival": "Defeated by Rival",
+  "defeated":"Defeated",
+  "luck":"Luck",
+  "score":"Score",
+  "mode":"Mode",
+  "challengeRules":"Rule(s)",
+  "challengeMonoGen1":"Gen I",
+  "challengeMonoGen2":"Gen II",
+  "challengeMonoGen3":"Gen III",
+  "challengeMonoGen4":"Gen IV",
+  "challengeMonoGen5":"Gen V",
+  "challengeMonoGen6":"Gen VI",
+  "challengeMonoGen7":"Gen VII",
+  "challengeMonoGen8":"Gen VIII",
+  "challengeMonoGen9":"Gen IX",
+  "playerItems":"Player Items",
+  "personalBest":"Personal Best!",
+  "SPDshortened":"Vel.",
+  "runInfo":"Run Info",
+  "money":"Money",
+  "runLength":"Run Length",
+  "viewHeldItems":"Held Items",
+  "hallofFameText":"Welcome to the Hall of Fame!",
+  "viewHallOfFame":"View Hall of Fame!"
```
```suggestion
  "victory": "胜利！",
  "defeatedWild": "被打败",
  "defeatedTrainer": "被打败",
  "defeatedTrainerDouble": "被组合打败",
  "defeatedRival": "被劲敌打败",
  "defeated":"被打败",
  "luck":"幸运",
  "score":"分数",
  "mode":"模式",
  "challengeRules":"规则",
  "challengeMonoGen1":"一代",
  "challengeMonoGen2":"二代",
  "challengeMonoGen3":"三代",
  "challengeMonoGen4":"四代",
  "challengeMonoGen5":"五代",
  "challengeMonoGen6":"六代",
  "challengeMonoGen7":"七代",
  "challengeMonoGen8":"八代",
  "challengeMonoGen9":"九代",
  "playerItems":"玩家道具",
  "personalBest":"个人最佳！",
  "SPDshortened":"Vel.",
  "runInfo":"游戏记录",
  "money":"金钱",
  "runLength":"游戏",
  "viewHeldItems":"持有道具",
  "hallofFameText":"欢迎来到名人堂！",
  "viewHallOfFame":"浏览名人堂！"
```

frutescens commented:
Hi, apologies. I uploaded some screenshots to the PR description. Thank you for your localization work.

Adri1 commented:
Questions :
How does this displays?
![image](https://github.com/user-attachments/assets/efbb1924-1249-4a44-8018-6a05ec6f063e)

It is possible to also be able have a translation entry for this? Because the money simple in languages like Frenc of German is located after the amount, not before :
![image](https://github.com/user-attachments/assets/b0995015-8062-467f-b099-99e4a532bc20)

Speaking of numbers do they dynamically adapt to the language format? As it's the case in most texts where numers appear, it detects the language used by the player's navigator (if I remeber well) to adapt it automatically, but idk if it's the case for your code

For example:
English : 1,234.56
French : 1 234,56

frutescens commented:
I am not sure. I use Pokerogue's own big number formatter so I will look into it + I will hopefully have a response for you in 12 or so hours.

NicusPulcis commented on src/locales/it/menu-ui-handler.ts:
```diff
@@ -4,6 +4,7 @@ export const menuUiHandler: SimpleTranslationEntries = {
   "GAME_SETTINGS": "Impostazioni",
   "ACHIEVEMENTS": "Obiettivi",
   "STATS": "Statistiche",
+  "RUN_HISTORY": "Run History",
```
```suggestion
  "RUN_HISTORY": "Run precedenti",
```

NicusPulcis commented on src/locales/it/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,35 @@
+import { SimpleTranslationEntries } from "#app/interfaces/locales";
+
+export const runHistory: SimpleTranslationEntries = {
+  "victory": "Victory!",
+  "defeatedWild": "Defeated by ",
+  "defeatedTrainer": "Defeated by ",
+  "defeatedTrainerDouble": "Defeated by Duo",
+  "defeatedRival": "Defeated by Rival",
+  "defeated":"Defeated",
+  "luck":"Luck",
+  "score":"Score",
+  "mode":"Mode",
+  "challengeRules":"Rule(s)",
+  "challengeMonoGen1":"Gen I",
+  "challengeMonoGen2":"Gen II",
+  "challengeMonoGen3":"Gen III",
+  "challengeMonoGen4":"Gen IV",
+  "challengeMonoGen5":"Gen V",
+  "challengeMonoGen6":"Gen VI",
+  "challengeMonoGen7":"Gen VII",
+  "challengeMonoGen8":"Gen VIII",
+  "challengeMonoGen9":"Gen IX",
+  "playerItems":"Player Items",
+  "personalBest":"Personal Best!",
+  "SPDshortened":"Vel.",
+  "runInfo":"Run Info",
+  "money":"Money",
+  "runLength":"Run Length",
+  "viewHeldItems":"Held Items",
+  "hallofFameText":"Welcome to the Hall of Fame!",
+  "viewHallOfFame":"View Hall of Fame!"
```
```suggestion
  "victory": "Vittoria!",
  "defeatedWild": "Sconfitti da ",
  "defeatedTrainer": "Sconfitti da ",
  "defeatedTrainerDouble": "Sconfitti dalla coppia ",
  "defeatedRival": "Sconfitti dal rivale",
  "defeated":"Sconfitti",
  "luck":"Fortuna",
  "score":"Punteggio",
  "mode":"Modalità",
  "challengeRules":"Regola/e",
  "challengeMonoGen1":"1ª gen",
  "challengeMonoGen2":"2ª gen",
  "challengeMonoGen3":"3ª gen",
  "challengeMonoGen4":"4ª gen",
  "challengeMonoGen5":"5ª gen",
  "challengeMonoGen6":"6ª gen",
  "challengeMonoGen7":"7ª gen",
  "challengeMonoGen8":"8ª gen",
  "challengeMonoGen9":"9ª gen",
  "playerItems":"Oggetti giocatore",
  "personalBest":"Record personale!",
  "SPDshortened":"Vel.",
  "runInfo":"Info Run",
  "money":"Patrimonio",
  "runLength":"Durata Run",
  "viewHeldItems":"Oggetti equipaggiati ",
  "hallofFameText":"Benvenuto alla Sala d'Onore!",
  "viewHallOfFame":"Visualizza Sala d'Onore!"
```

NicusPulcis reviewed (approved):
Italian has been localized

CodeTappert commented:
I am almost certain that the german text is too long

Adri1 commented on src/locales/fr/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,35 @@
+  "defeatedTrainerDouble": "Battu par Duo",
+  "defeatedRival": "Battu par Rival",
+  "defeated":"Vaincu",
+  "luck":"Chance",
+  "score":"Score",
+  "mode":"Mode",
+  "challengeRules":"Règle·s",
+  "challengeMonoGen1":"1G",
+  "challengeMonoGen2":"2G",
+  "challengeMonoGen3":"3G",
+  "challengeMonoGen4":"4G",
+  "challengeMonoGen5":"5G",
+  "challengeMonoGen6":"6G",
+  "challengeMonoGen7":"7G",
+  "challengeMonoGen8":"8G",
+  "challengeMonoGen9":"9G",
+  "playerItems":"Objets joueur",
+  "personalBest":"Record personnel !",
+  "SPDshortened":"Vit.",
+  "runInfo":"Infos session",
+  "money":"Argent",
+  "runLength":"Durée session",
```
```suggestion
  "luck":"Chance ",
  "score":"Score",
  "mode":"Mode",
  "challengeRules":"Règles ",
  "challengeMonoGen1":"1G",
  "challengeMonoGen2":"2G",
  "challengeMonoGen3":"3G",
  "challengeMonoGen4":"4G",
  "challengeMonoGen5":"5G",
  "challengeMonoGen6":"6G",
  "challengeMonoGen7":"7G",
  "challengeMonoGen8":"8G",
  "challengeMonoGen9":"9G",
  "playerItems":"Objets joueur",
  "personalBest":"Record personnel !",
  "SPDshortened":"Vit.",
  "runInfo":"Infos session",
  "money":"Argent",
  "runLength":"Durée session ",
```

Adri1 reviewed (commented):
First French corrections

CodeTappert commented:
Uploading 2024-08-02 19-24-38.mp4…

Found a Bug.

CodeTappert commented:
![image](https://github.com/user-attachments/assets/90afc1bd-bee0-4691-96ba-c33f6ca9555c)
That background also looks not correct

CodeTappert commented on src/locales/de/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,35 @@
+  "SPDshortened":"Geschw.",
+  "runInfo":"Durchlauf Info.",
+  "money":"Geld",
+  "runLength":"Länge des Durchlaufs",
```
```suggestion
  "runLength":"Durchlauf Dauer",
```

CodeTappert commented on src/locales/de/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,35 @@
+  "playerItems":"Spielergegenstände",
+  "personalBest":"Persönlicher Bestwert!",
+  "SPDshortened":"Geschw.",
+  "runInfo":"Durchlauf Info.",
```
```suggestion
  "runInfo":"Durchlauf Informationen",
```

CodeTappert commented on src/locales/de/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,35 @@
+  "challengeMonoGen9":"Gen IX",
+  "playerItems":"Spielergegenstände",
+  "personalBest":"Persönlicher Bestwert!",
+  "SPDshortened":"Geschw.",
```
Where is this used?

frutescens replied:
SPDshortened is only used in Spanish/Portuguese because Veloc. is too long

frutescens commented:
> 2024-08-02.19-24-38.mp4 
> Found a Bug.
> 
> Only option then is to reload

Fixed! Thank you for letting me know. I also replaced the hall of fame image.

Adri1 commented on src/locales/fr/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,35 @@
+  "challengeMonoGen7":"7G",
+  "challengeMonoGen8":"8G",
+  "challengeMonoGen9":"9G",
+  "playerItems":"Objets joueur",
```
```suggestion
  "playerItems":"Objets Dresseur",
```

Adri1 reviewed (commented):
French adjustment

frutescens commented:
![image](https://github.com/user-attachments/assets/8023d567-30d6-4fa5-9576-5c0dcd868f87)

frutescens commented:
![image](https://github.com/user-attachments/assets/33747bb0-4a68-4df1-a4f5-bf9d01046757)
![image](https://github.com/user-attachments/assets/3fcd2c27-b80e-4c20-8bb3-561e80034ccb)

frutescens commented:
![image](https://github.com/user-attachments/assets/ccd77afe-6677-4499-9fde-f494392b1fa7)

Adri1 commented on src/locales/fr/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,36 @@
+  "runLength":"Durée session ",
+  "viewHeldItems":"Objets tenus",
+  "hallofFameText":"Bienvenue au Panthéon !",
+  "viewHallOfFame":"Voir le Panthéon !",
+  "viewEndingSplash":"View ending art!"
```
```suggestion
  "viewHallOfFame":"Voir le Panthéon",
  "viewEndingSplash":"Voir l’illustration de fin !"
```

Adri1 commented on src/locales/fr/menu-ui-handler.ts:
```diff
@@ -18,10 +19,12 @@ export const menuUiHandler: SimpleTranslationEntries = {
-  "unlinkDiscord": "Délier Discord",
-  "linkGoogle": "Lier à Google",
-  "unlinkGoogle": "Délier Google",
+  "importRunHistory":"Import Run History",
+  "exportRunHistory":"Export Run History",
+  "linkDiscord": "Link Discord",
+  "unlinkDiscord": "Unlink Discord",
+  "linkGoogle": "Link Google",
+  "unlinkGoogle": "Unlink Google",
```
```suggestion
  "importRunHistory":"Importer historique",
  "exportRunHistory":"Exporter historique",
  "linkDiscord": "Lier à Discord",
  "unlinkDiscord": "Délier Discord",
  "linkGoogle": "Lier à Google",
  "unlinkGoogle": "Délier Google",
```

DustinLin commented on src/system/game-data.ts:
```diff
@@ -551,6 +568,84 @@ export class GameData {
+        const cachedRHData = cachedResponse ?? {};
+        //check to see whether cachedData or serverData is more up-to-date
+        /**
+       * Networking Code
```
if local storage is temporary, maybe make a note of it here in the comments, maybe something like "Networking Code (below code for eventual? move to online storage")?

DayKev commented on src/ui/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,338 @@
+        const gameOutcomeLabel = addTextObject(this.scene, 0, 0, `${i18next.t("runHistory:defeatedWild")}`, TextStyle.WINDOW);
+        enemyContainer.add(gameOutcomeLabel);
+        data.enemyParty.forEach((enemyData, e) => {
+          //This allows the enemyParty to be shown - doubles or sings -> 58+(e*8)
```
```suggestion
          // This allows the enemyParty to be shown - doubles or singles -> 58+(e*8)
```
Presumably you meant "singles" and not "sings"? 🤔

DayKev commented on src/ui/run-info-ui-handler.ts:
```diff
@@ -0,0 +1,722 @@
+  // The use of non-directional / A / B buttons is named in relation to functions used during starter-select.
+  // Button.CYCLE_FORM (F key) --> displays ending art (victory only)
+  // Button.CYCLE_SHINY (R key) --> displays hall of fame (victory only)
+  // Button.CYCLE_ABILITY (E key) --> shows pokemon held items
```
Should probably be tsdocs?

DayKev commented on src/ui/run-info-ui-handler.ts:
```diff
@@ -0,0 +1,722 @@
+  createHallofFame(): void {
+    //Issue Note (08-05-2024): It seems as if fused pokemon do not appear with the averaged color b/c pokemonData's loadAsset requires there to be some active battle?
+    this.hallofFameContainer = this.scene.add.container(0, 0);
+    //Thank you Hayuna for the code
```
```suggestion
    // Thank you Hayuna for the code
```

DayKev commented on src/ui/run-info-ui-handler.ts:
```diff
@@ -0,0 +1,722 @@
+
+
+  createHallofFame(): void {
+    //Issue Note (08-05-2024): It seems as if fused pokemon do not appear with the averaged color b/c pokemonData's loadAsset requires there to be some active battle?
```
```suggestion
    // Issue Note (08-05-2024): It seems as if fused pokemon do not appear with the averaged color b/c pokemonData's loadAsset requires there to be some active battle?
```

DayKev commented on src/ui/run-info-ui-handler.ts:
```diff
@@ -0,0 +1,722 @@
+      const speedLabel = (currentLanguage==="es"||currentLanguage==="pt_BR") ? i18next.t("runHistory:SPDshortened") : i18next.t("pokemonInfo:Stat.SPDshortened");
+      const speed = speedLabel+": "+pStats[5];
+
+      //Column 1: HP Atk Def
```
```suggestion
      // Column 1: HP Atk Def
```

DayKev commented on src/ui/run-info-ui-handler.ts:
```diff
@@ -0,0 +1,722 @@
+      pokeStatText1.appendText(atk);
+      pokeStatText1.appendText(def);
+      pokeStatTextContainer.add(pokeStatText1);
+      //Column 2: SpAtk SpDef Speed
```
```suggestion
      // Column 2: SpAtk SpDef Speed
```

DayKev commented on src/ui/run-info-ui-handler.ts:
```diff
@@ -0,0 +1,722 @@
+          enemyContainer.setPosition(8, 14);
+          break;
+        }
+      //Trainer - Single and Double
```
```suggestion
      // Trainer - Single and Double
```

DayKev commented on src/ui/run-info-ui-handler.ts:
```diff
@@ -0,0 +1,722 @@
+      if (runData.battleType === BattleType.WILD) {
+        switch (runData.enemyParty.length) {
+        case 1:
+          //Wild - Singles
```
```suggestion
          // Wild - Singles
```

DayKev commented on src/ui/run-info-ui-handler.ts:
```diff
@@ -0,0 +1,722 @@
+    if (!runResult) {
+      const enemyContainer = this.scene.add.container(0, 0);
+
+      //Wild - Single and Doubles
```
```suggestion
      // Wild - Single and Doubles
```

DayKev commented on src/ui/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,338 @@
+        const tObj = data.trainer.toTrainer(this.scene);
+        if (data.trainer.trainerType >= 375) {
+          const gameOutcomeLabel = addTextObject(this.scene, 8, 5, `${i18next.t("runHistory:defeatedRival")}`, TextStyle.WINDOW);
+          //otherwise it becomes Rival_5 in Ivy's case
```
```suggestion
          // otherwise it becomes Rival_5 in Ivy's case
```

DayKev commented on src/system/game-data.ts:
```diff
@@ -551,6 +568,84 @@ export class GameData {
+    runHistoryData[timestamp] = {};
+    runHistoryData[timestamp]["entry"] = runEntry;
+    runHistoryData[timestamp]["victory"] = victory;
+    //Not implemented at the moment, but leaving room for future work
```
```suggestion
    // Not implemented at the moment, but leaving room for future work
```

DayKev commented on src/system/game-data.ts:
```diff
@@ -551,6 +568,84 @@ export class GameData {
+    const timestamps = Object.keys(runHistoryData);
+    const timestampsNo = timestamps.map(Number);
+
+    //Arbitrary limit of 25 entries per User --> Can increase or decrease
```
```suggestion
    // Arbitrary limit of 25 entries per User --> Can increase or decrease
```

DayKev commented on src/system/game-data.ts:
```diff
@@ -551,6 +568,84 @@ export class GameData {
+          cachedResponse = JSON.parse(decrypt(cachedResponse, true));
+        }
+        const cachedRHData = cachedResponse ?? {};
+        //check to see whether cachedData or serverData is more up-to-date
```
```suggestion
        // check to see whether cachedData or serverData is more up-to-date
```

DayKev commented on src/locales/zh_TW/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,36 @@
+  "viewEndingSplash":"View ending art!"
+} as const;
+
+//Mode Information found in game-mode.ts
+//Wave / Lv found in save-slot-select-ui-handler.ts
```
```suggestion
// Mode Information found in game-mode.ts
// Wave / Lv found in save-slot-select-ui-handler.ts
```

DayKev commented on src/locales/zh_CN/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,36 @@
+  "viewEndingSplash":"View ending art!"
+} as const;
+
+//Mode Information found in game-mode.ts
+//Wave / Lv found in save-slot-select-ui-handler.ts
```
```suggestion
// Mode Information found in game-mode.ts
// Wave / Lv found in save-slot-select-ui-handler.ts
```

DayKev commented on src/locales/pt_BR/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,36 @@
+  "viewEndingSplash":"View ending art!"
+} as const;
+
+//Mode Information found in game-mode.ts
+//Wave / Lv found in save-slot-select-ui-handler.ts
```
```suggestion
// Mode Information found in game-mode.ts
// Wave / Lv found in save-slot-select-ui-handler.ts
```

DayKev commented on src/locales/ko/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,36 @@
+  "viewEndingSplash":"View ending art!"
+} as const;
+
+//Mode Information found in game-mode.ts
+//Wave / Lv found in save-slot-select-ui-handler.ts
```
```suggestion
// Mode Information found in game-mode.ts
// Wave / Lv found in save-slot-select-ui-handler.ts
```

DayKev commented on src/locales/ja/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,36 @@
+  "viewEndingSplash":"View ending art!"
+} as const;
+
+//Mode Information found in game-mode.ts
+//Wave / Lv found in save-slot-select-ui-handler.ts
```
```suggestion
// Mode Information found in game-mode.ts
// Wave / Lv found in save-slot-select-ui-handler.ts
```

DayKev commented on src/locales/it/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,36 @@
+  "viewEndingSplash":"View ending art!"
+} as const;
+
+//Mode Information found in game-mode.ts
+//Wave / Lv found in save-slot-select-ui-handler.ts
```
```suggestion
// Mode Information found in game-mode.ts
// Wave / Lv found in save-slot-select-ui-handler.ts
```

DayKev commented on src/locales/fr/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,36 @@
+  "viewEndingSplash":"Voir l’illustration de fin !"
+} as const;
+
+//Mode Information found in game-mode.ts
+//Wave / Lv found in save-slot-select-ui-handler.ts
```
```suggestion
// Mode Information found in game-mode.ts
// Wave / Lv found in save-slot-select-ui-handler.ts
```

DayKev commented on src/locales/es/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,36 @@
+  "viewEndingSplash":"View ending art!"
+} as const;
+
+//Mode Information found in game-mode.ts
+//Wave / Lv found in save-slot-select-ui-handler.ts
```
```suggestion
// Mode Information found in game-mode.ts
// Wave / Lv found in save-slot-select-ui-handler.ts
```

DayKev commented on src/locales/en/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,36 @@
+  "viewEndingSplash":"View ending art!"
+} as const;
+
+//Mode Information found in game-mode.ts
+//Wave / Lv found in save-slot-select-ui-handler.ts
```
```suggestion
// Mode Information found in game-mode.ts
// Wave / Lv found in save-slot-select-ui-handler.ts
```

DayKev commented on src/locales/de/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,36 @@
+  "viewEndingSplash":"View ending art!"
+} as const;
+
+//Mode Information found in game-mode.ts
+//Wave / Lv found in save-slot-select-ui-handler.ts
```
```suggestion
// Mode Information found in game-mode.ts
// Wave / Lv found in save-slot-select-ui-handler.ts
```

DayKev commented on src/locales/ca-ES/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,36 @@
+  "viewEndingSplash":"View ending art!"
+} as const;
+
+//Mode Information found in game-mode.ts
+//Wave / Lv found in save-slot-select-ui-handler.ts
```
```suggestion
// Mode Information found in game-mode.ts
// Wave / Lv found in save-slot-select-ui-handler.ts
```

DayKev reviewed (changes requested):
Hello, the style police is here. 🧐

AsdarDevelops commented on src/locales/es/menu-ui-handler.ts:
```diff
@@ -4,6 +4,7 @@ export const menuUiHandler: SimpleTranslationEntries = {
   "GAME_SETTINGS": "Ajustes",
   "ACHIEVEMENTS": "Logros",
   "STATS": "Estadísticas",
+  "RUN_HISTORY": "Run History",
```
```suggestion
  "RUN_HISTORY": "Historial de partida",
```

AsdarDevelops commented on src/locales/es/menu-ui-handler.ts:
```diff
@@ -16,6 +17,8 @@ export const menuUiHandler: SimpleTranslationEntries = {
   "importSlotSelect": "Selecciona una ranura para importar.",
   "exportSession": "Exportar Sesión",
   "exportSlotSelect": "Selecciona una ranura para exportar.",
+  "importRunHistory":"Import Run History",
+  "exportRunHistory":"Export Run History",
```
```suggestion
  "importRunHistory":"Importar Historial de partida",
  "exportRunHistory":"Exportar Historial de partida",
```

AsdarDevelops commented on src/locales/es/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,36 @@
+import { SimpleTranslationEntries } from "#app/interfaces/locales";
+
+export const runHistory: SimpleTranslationEntries = {
+  "victory": "Victory!",
+  "defeatedWild": "Defeated by ",
+  "defeatedTrainer": "Defeated by ",
+  "defeatedTrainerDouble": "Defeated by Duo",
+  "defeatedRival": "Defeated by Rival",
+  "defeated":"Defeated",
+  "luck":"Luck",
+  "score":"Score",
+  "mode":"Mode",
+  "challengeRules":"Rule(s)",
+  "challengeMonoGen1":"Gen I",
+  "challengeMonoGen2":"Gen II",
+  "challengeMonoGen3":"Gen III",
+  "challengeMonoGen4":"Gen IV",
+  "challengeMonoGen5":"Gen V",
+  "challengeMonoGen6":"Gen VI",
+  "challengeMonoGen7":"Gen VII",
+  "challengeMonoGen8":"Gen VIII",
+  "challengeMonoGen9":"Gen IX",
+  "playerItems":"Player Items",
+  "personalBest":"Personal Best!",
+  "SPDshortened":"Vel.",
+  "runInfo":"Run Info",
+  "money":"Money",
+  "runLength":"Run Length",
+  "viewHeldItems":"Held Items",
+  "hallofFameText":"Welcome to the Hall of Fame!",
+  "viewHallOfFame":"View Hall of Fame!",
+  "viewEndingSplash":"View ending art!"
```
```suggestion
  "victory": "Victory!",
  "defeatedWild": "Derrotado por ",
  "defeatedTrainer": "Derrotado por ",
  "defeatedTrainerDouble": "Derrotado por un dúo",
  "defeatedRival": "Derrotado por el rival",
  "defeated":"Derrotado",
  "luck":"Suerte",
  "score":"Puntuación",
  "mode":"Modo",
  "challengeRules":"Regla(s)",
  "challengeMonoGen1":"Gen I",
  "challengeMonoGen2":"Gen II",
  "challengeMonoGen3":"Gen III",
  "challengeMonoGen4":"Gen IV",
  "challengeMonoGen5":"Gen V",
  "challengeMonoGen6":"Gen VI",
  "challengeMonoGen7":"Gen VII",
  "challengeMonoGen8":"Gen VIII",
  "challengeMonoGen9":"Gen IX",
  "playerItems":"Objetos del jugador",
  "personalBest":"¡Récord personal!",
  "SPDshortened":"Vel.",
  "runInfo":"Info. de partida",
  "money":"Dinero",
  "runLength":"Duración de partida",
  "viewHeldItems":"Objetos equipados",
  "hallofFameText":"¡Bienvenido al Hall de la Fama!",
  "viewHallOfFame":"¡Ver Hall de la Fama!",
  "viewEndingSplash":"¡Ver la imagen final!"
```

AsdarDevelops reviewed (commented):
Spanish done

flx-sta commented on src/loading-scene.ts:
```diff
@@ -215,6 +215,10 @@ export class LoadingScene extends SceneBase {
     this.loadAtlas("c_rival_m", "character", "rival_m");
     this.loadAtlas("c_rival_f", "character", "rival_f");
 
+    //Load run-history related images
+    this.loadImage("hall_of_fame_red", "ui");
+    this.loadImage("hall_of_fame_blue", "ui");
+
```
do we want to load those immediately?
Can't we just load these when someone opens the UI?

_I know this has always been done that way but I'm really not a fan of stuffing EVERYTHING possible into the memory at the beginning of the game_

frutescens replied:
I can try looking into it, but it seems like you need access to class LoadingScene / SceneBase to load an image. I'll try to find another way though.

flx-sta replied:
@frutescens `BattleScene` is also an extension of `SceneBase` and almost everything has access to `BattleSccene`

flx-sta replied:
1 thing to watch out for is that `this.loadImage()` is an async action so you would have to make sure the image is loaded

frutescens replied:
Tested on slow 4G ---> hopefully this means it's ok

flx-sta replied:
Should be I guess. Worst things worst it’s an image missing

frutescens replied:
true, and there's the ending splash image below the hall of fame images anyways. thank you for your help.

flx-sta commented on src/phases.ts:
```diff
@@ -4313,6 +4313,13 @@ export class GameOverPhase extends BattlePhase {
             this.scene.gameData.gameStats.dailyRunSessionsWon++;
           }
         }
+        this.scene.gameData.getSession(this.scene.sessionSlotId).then(sessionData => {
+          if (sessionData) {
+            this.scene.gameData.saveRunHistory(this.scene, sessionData, this.victory);
+          }
+        }).catch(err => {
+          console.error(err);
+        });
```
1. I assume this is meant to run asynchronously!?
2. Please add some context to the `console.error` (like `console.error("Failed to save run to history: ", err)`)

flx-sta commented on src/system/game-data.ts:
```diff
@@ -181,6 +183,16 @@ export const AbilityAttr = {
   ABILITY_HIDDEN: 4
 };
 
+export interface RunHistoryData {
+  [key: integer]: RunEntries;
+}
```
Can be simplified and please don't add any more `integer`:
```suggestion
export type RunHistoryData  = Record<number, RunEntry> 
```
<sub> Note that I used singular `RunEntry` here. See next comment </sub>

flx-sta commented on src/system/game-data.ts:
```diff
@@ -181,6 +183,16 @@ export const AbilityAttr = {
+  [key: integer]: RunEntries;
+}
+
+export interface RunEntries {
+  entry: SessionSaveData;
+  victory: boolean;
+  favorite: boolean;
+}
```
This should be singular. The plural indicates a list/array but this isn't one

```suggestion
export interface RunEntry {
  entry: SessionSaveData;
  victory: boolean;
  favorite: boolean;
}
```

flx-sta commented on src/system/game-data.ts:
```diff
@@ -442,6 +456,9 @@ export class GameData {
         if (versions[0] !== versions[1]) {
           const [ versionNumbers, oldVersionNumbers ] = versions.map(ver => ver.split('.').map(v => parseInt(v)));
         }*/
+        if (!localStorage.hasOwnProperty(`runHistoryData_${loggedInUser?.username}`)) {
+          localStorage.setItem(`runHistoryData_${loggedInUser?.username}`, encrypt("", true));
+        }
```
Please don't use `hasOwnProperty()`. It's not recommended 
<sub>can't find the source for that but below is a ChatGPT response</sub>

Instead do this
```suggestion
        const lsItemKey = `runHistoryData_${loggedInUser?.username}`
        const lsItem = localStorage.getItem(lsItemKey);
        if (!lsItem) {
          localStorage.setItem(lsItemKey, encrypt("", true));
        }
```

----

> Using `hasOwnProperty` on `localStorage` is not recommended because `localStorage` is an instance of the `Storage` interface, which is a key-value store for string data. The `Storage` object doesn't directly inherit properties from `Object.prototype`, so methods like `hasOwnProperty` might not behave as expected.
> 
> Instead, you can check for the existence of a key in `localStorage` using the `localStorage.getItem(key) !== null` pattern, which is more reliable. Here's an example:
> 
> ```javascript
> if (localStorage.getItem('yourKey') !== null) {
>   // Key exists in localStorage
> } else {
>   // Key does not exist in localStorage
> }
> ```
> 
> This approach is safer and aligns with the intended usage of the `Storage` API.

flx-sta commented on src/system/game-data.ts:
```diff
@@ -553,6 +570,84 @@ export class GameData {
     });
   }
 
+  public async getRunHistoryData(scene: BattleScene): Promise<Object> {
```
using `Object` as a response type is discouraged as it's very unspecific and close to using `any`.

Instead return the specific type expected. 
You can type cast the `JSON.parse` => 
```ts
JSON.parse(decrypt(cachedResponse, true)) as RunHistoryData;
```
<sub>if that's what's expected</sub>

flx-sta commented on src/system/game-data.ts:
```diff
@@ -553,6 +570,84 @@ export class GameData {
+      const response = await Utils.apiFetch("savedata/runHistory", true);
+      const data = await response.json();
+      */
+      if (localStorage.hasOwnProperty(`runHistoryData_${loggedInUser?.username}`)) {
+        let cachedResponse = localStorage.getItem(`runHistoryData_${loggedInUser?.username}`);
```
Please don't use `hasOwnProperty()`. It's not recommended 
<sub>can't find the source for that but below is a ChatGPT response</sub>

Instead do this
```suggestion
      const lsItemKey = `runHistoryData_${loggedInUser?.username}`;
      const lsItem = localStorage.getItem(lsItemKey);
      if (lsItem) {
        let cachedResponse = lsItem;
```
----

> Using `hasOwnProperty` on `localStorage` is not recommended because `localStorage` is an instance of the `Storage` interface, which is a key-value store for string data. The `Storage` object doesn't directly inherit properties from `Object.prototype`, so methods like `hasOwnProperty` might not behave as expected.
> 
> Instead, you can check for the existence of a key in `localStorage` using the `localStorage.getItem(key) !== null` pattern, which is more reliable. Here's an example:
> 
> ```javascript
> if (localStorage.getItem('yourKey') !== null) {
>   // Key exists in localStorage
> } else {
>   // Key does not exist in localStorage
> }
> ```
> 
> This approach is safer and aligns with the intended usage of the `Storage` API.

flx-sta commented on src/system/game-data.ts:
```diff
@@ -553,6 +570,84 @@ export class GameData {
+
+    localStorage.setItem(`runHistoryData_${loggedInUser?.username}`, encrypt(JSON.stringify(runHistoryData), true));
+
+    /**
+     * Networking Code DO NOT DELETE
+     *
+    if (!Utils.isLocal) {
+      try {
+        Utils.apiPost("savedata/runHistory", JSON.stringify(runHistoryData), undefined, true);
+        return true;
+      } catch (err) {
+        console.log("savedata/runHistory POST failed : ", err);
+        return false;
+      }
+    }
+    */
```
the networking code is not using `async/await`

flx-sta commented on src/system/game-data.ts:
```diff
@@ -553,6 +570,84 @@ export class GameData {
+    }
+
+    const timestamp = (runEntry.timestamp).toString();
+    runHistoryData[timestamp] = {};
+    runHistoryData[timestamp]["entry"] = runEntry;
+    runHistoryData[timestamp]["victory"] = victory;
+    // Not implemented at the moment, but leaving room for future work
+    runHistoryData[timestamp]["favorite"] = false;
```
```suggestion
    runHistoryData[timestamp] = {
      entry: runEntry,
      victory: victory,
    };
```

And you should move the `Not implemented at the moment, but leaving room for future work` into the description of the interface

```ts
export interface RunEntry {
  entry: SessionSaveData;
  victory: boolean;
  /** Not used yet. Potential future feature */
  favorite?: boolean;
}

flx-sta commented on src/system/game-data.ts:
```diff
@@ -553,6 +570,84 @@ export class GameData {
+    }
+  }
+
+  async saveRunHistory(scene: BattleScene, runEntry : SessionSaveData, victory: boolean): Promise<boolean> {
```
```suggestion
  async saveRunHistory(scene: BattleScene, runEntry : SessionSaveData, isVictory: boolean): Promise<boolean> {
```

to make it more clear that `victory` is a boolean

flx-sta commented on src/ui/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,338 @@
+    } else if (error) {
+      ui.playError();
+    }
+    return success || error;
```
shouldn't it be `success ?? error`?
Or should `success=false` also roll back to `error`?

frutescens replied:
Should be the latter. But, I did take this code from older UI code which is you know, flawed.

flx-sta replied:
resolved I guess

flx-sta commented on src/ui/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,338 @@
+    }
+  }
+
+  showText(text: string, delay?: integer, callback?: Function, callbackDelay?: integer, prompt?: boolean, promptDelay?: integer) {
```
no more `integer` please

```suggestion
  showText(text: string, delay?: number, callback?: Function, callbackDelay?: number, prompt?: boolean, promptDelay?: number) {
```

flx-sta commented on src/ui/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,338 @@
+    this.runSelectMessageBoxContainer.setVisible(!!text?.length);
+  }
+
+  setCursor(cursor: integer): boolean {
```
no more `integer` please

```suggestion
  setCursor(cursor: number): boolean {
```

flx-sta commented on src/ui/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,338 @@
+    return changed;
+  }
+
+  setScrollCursor(scrollCursor: integer): boolean {
```
no more `integer`:

```suggestion
  setScrollCursor(scrollCursor: number): boolean {
```

flx-sta commented on src/ui/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,338 @@
+}
+
+class RunEntry extends Phaser.GameObjects.Container {
+  public slotId: integer;
```
no more `integer`:
```suggestion
  public slotId: number;
```

flx-sta commented on src/ui/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,338 @@
+  public entryData: RunHistoryData;
+  private loadingLabel: Phaser.GameObjects.Text;
+
+  constructor(scene: BattleScene, runHistory: any, timestamp: string, slotId: integer) {
```
no more `integer`
```suggestion
  constructor(scene: BattleScene, runHistory: any, timestamp: string, slotId: number) {
```

frutescens commented:
I forgot to fix the cursor issue today.... nooooo.... tomorrow.

flx-sta commented on src/ui/run-info-ui-handler.ts:
```diff
@@ -194,7 +192,9 @@ export default class GameInfoUiHandler extends UiHandler {
         const tObjSpriteKey = tObj.config.getSpriteKey(this.runInfo.trainer.variant === TrainerVariant.FEMALE, false);
+        //const tObjSprite = this.scene.add.image(2, 10, "rival_f");
         const tObjSprite = this.scene.add.sprite(0, 5, tObjSpriteKey);
+        console.log(tObjSpriteKey);
```
debugging leftovers?
```suggestion
        const tObjSprite = this.scene.add.sprite(0, 5, tObjSpriteKey);
```

flx-sta commented on src/ui/run-info-ui-handler.ts:
```diff
@@ -391,12 +391,11 @@ export default class GameInfoUiHandler extends UiHandler {
       const passiveLabel = (currentLanguage==="ko"||currentLanguage==="zh_CN"||currentLanguage==="zh_TW") ? i18next.t("starterSelectUiHandler:passive") : (i18next.t("starterSelectUiHandler:passive") as string).charAt(0);
       const abilityLabel = (currentLanguage==="ko"||currentLanguage==="zh_CN"||currentLanguage==="zh_TW") ? i18next.t("starterSelectUiHandler:ability") : (i18next.t("starterSelectUiHandler:ability") as string).charAt(0);
-      const pPassiveInfo = pokemon.passive ? `${passiveLabel+": "+allAbilities[starterPassiveAbilities[pSpecies.speciesId]].name}` : "";
+      const pPassiveInfo = pokemon.passive ? `${passiveLabel+": "+pokemon.getPassiveAbility().name}` : "";
```
you are making it harder for yourself here. Use a template string to it's full potential:
```suggestion
      const pPassiveInfo = pokemon.passive ? `${passiveLabel}: ${pokemon.getPassiveAbility().name}` : "";
```

Adri1 reviewed (commented):
Tiny French correction

Corrade commented on src/ui/run-info-ui-handler.ts:
```diff
@@ -0,0 +1,720 @@
+    let success = false;
+    const error = false;
+
+    if (button === Button.CANCEL) {
```
This could be moved into the switch statement below.

Corrade commented on src/ui/run-info-ui-handler.ts:
```diff
@@ -0,0 +1,720 @@
+      break;
+    case Button.CYCLE_SHINY:
+      if (this.isVictory) {
+        if (!this.hallofFameContainer.visible) {
```
This could be simplified to this.hallofFameContainer.setVisible(!this.hallofFameContainer.visible). This conveys the toggling behaviour in one line.

Similar comment for party visibility below.

frutescens replied:
This causes an 'attempting to setVisible() an undefined object' because this.hallofFameContainer is not created for a loser's run.

Corrade reviewed (changes requested):
Thanks a lot for this feature. I hope my comments are useful!

flx-sta commented on src/ui/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,315 @@
+    let success = false;
+    const error = false;
+
+    if (button === Button.ACTION || button === Button.CANCEL) {
```
I disagree. 
The usage for switch-cases is usually fairly niche and was generally over-used in this project.

There is room for improvements though. Like this:
```suggestion
    if ([Button.ACTION, Button.CANCEL].includes(button)) {
```

frutescens commented on src/ui/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,315 @@
+        this.add(enemyContainer);
+      } else if (data.battleType === BattleType.TRAINER) {
+        const tObj = data.trainer.toTrainer(this.scene);
+        if (data.trainer.trainerType >= 375) {
```
That's the ID of the rival trainers, which have some funky code behind it. But things may have changed between then and now so I should double check.

flx-sta replied:
Then give it the name `const RIVAL_TRAINER_ID_THRESHHOLD =375`

frutescens replied:
yes, I just wanted to double-check that the addition of new trainers did not change its # in enum/trainer-types.ts

frutescens commented on src/ui/run-info-ui-handler.ts:
```diff
@@ -0,0 +1,720 @@
+      modeText.appendText(`\t\t${i18next.t("runHistory:challengeRules")}: `);
+      const runChallenges = this.runInfo.challenges;
+      const rules = [];
+      for (let i = 0; i < runChallenges.length; i++) {
```
Yes, unfortunately.... it is what it is. Maybe after this, I can try fixing up how challenges are stored. I don't get why non-activated challenges are stored with activated challenges....

frutescens commented on src/ui/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,315 @@
+      const gameOutcomeLabel = addTextObject(this.scene, 8, 5, `${i18next.t("runHistory:victory")}`, TextStyle.WINDOW);
+      this.add(gameOutcomeLabel);
+    } else {
+      if (data.battleType === BattleType.WILD) {
```
I put the else there to indicate that the difference was between this.victory and !this.victory.

frutescens commented on src/ui/run-info-ui-handler.ts:
```diff
@@ -0,0 +1,720 @@
+    super(scene, Mode.RUN_INFO);
+  }
+
+  	async setup() {
```
I'm using tabs though..

frutescens commented on src/ui/run-info-ui-handler.ts:
```diff
@@ -0,0 +1,720 @@
+      // Wild - Single and Doubles
+      if (this.runInfo.battleType === BattleType.WILD) {
+        switch (this.runInfo.enemyParty.length) {
+        case 1:
```
Hi, I get that but I wanted to make it easier for people to read and position the sprites. Yes, it is unnecessary code and can be simplified but I think it is useful for centering the sprite or sprites.

frutescens commented:
![image](https://github.com/user-attachments/assets/a0293f0c-b576-4953-a927-411e3604241a)

Adri1 commented on src/locales/fr/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,36 @@
+  "viewHeldItems":"Objets tenus",
+  "hallofFameText":"Bienvenue au Panthéon !",
+  "viewHallOfFame":"Voir le Panthéon",
+  "viewEndingSplash":"Voir l’illustration de fin !"
```
```suggestion
  "viewEndingSplash":"Voir l’illustration de fin"
```

Adri1 reviewed (commented):
Tiny correction French

sodaMelon commented on src/locales/ko/menu-ui-handler.ts:
```diff
@@ -16,6 +17,8 @@ export const menuUiHandler: SimpleTranslationEntries = {
   "importSlotSelect": "불러올 슬롯을 골라주세요.",
   "exportSession": "세션 내보내기",
   "exportSlotSelect": "내보낼 슬롯을 골라주세요.",
+  "importRunHistory":"Import Run History",
+  "exportRunHistory":"Export Run History",
```
```suggestion
  "importRunHistory":"플레이 이력 불러오기",
  "exportRunHistory":"플레이 이력 내보내기",
```

EnochG1 commented:
@frutescens 
Can you show us screenshot to check this label. (English is ok.)
```Typescript
  "viewHallOfFame":"View Hall of Fame!",
  "viewEndingSplash":"View ending art!"
```

josericardo-fo commented on src/locales/pt_BR/menu-ui-handler.ts:
```diff
@@ -16,6 +17,8 @@ export const menuUiHandler: SimpleTranslationEntries = {
   "importSlotSelect": "Selecione um slot para importar.",
   "exportSession": "Exportar sessão",
   "exportSlotSelect": "Selecione um slot para exportar.",
+  "importRunHistory":"Import Run History",
+  "exportRunHistory":"Export Run History",
```
```suggestion
  "importRunHistory":"Importar Histórico de Jogos",
  "exportRunHistory":"Exportar Histórico de Jogos",
```

josericardo-fo commented on src/locales/pt_BR/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,36 @@
+  "viewHeldItems": "Itens Segurados",
+  "hallofFameText": "Bem-vindo(a) ao Hall da Fama!",
+  "viewHallOfFame": "Veja o Hall da Fama!",
+  "viewEndingSplash":"View ending art!"
```
```suggestion
  "viewEndingSplash":"Veja a arte final!"
```

CodeTappert commented on src/locales/de/menu-ui-handler.ts:
```diff
@@ -16,6 +17,8 @@ export const menuUiHandler: SimpleTranslationEntries = {
   "importSlotSelect": "Wähle einen Slot zum Importieren.",
   "exportSession": "Sitzung exportieren",
   "exportSlotSelect": "Wähle einen Slot zum Exportieren.",
+  "importRunHistory":"Import Run History",
+  "exportRunHistory":"Export Run History",
```
```suggestion
  "importRunHistory":"Laufhistorie importieren",
  "exportRunHistory":"Laufhistorie exportieren",
```

CodeTappert commented on src/locales/de/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,36 @@
+  "viewHeldItems":"Getragene Items",
+  "hallofFameText":"Willkommen in der Ruhmeshalle",
+  "viewHallOfFame":"Ruhmeshalle ansehen!",
+  "viewEndingSplash":"View ending art!"
```
```suggestion
  "viewEndingSplash":"Endgrafik anzeigen!"
```

NicusPulcis commented on src/locales/it/menu-ui-handler.ts:
```diff
@@ -16,6 +17,8 @@ export const menuUiHandler: SimpleTranslationEntries = {
   "importSlotSelect": "Seleziona uno slot in cui importare.",
   "exportSession": "Esporta sessione",
   "exportSlotSelect": "Seleziona uno slot da cui esportare.",
+  "importRunHistory":"Import Run History",
+  "exportRunHistory":"Export Run History",
```
```suggestion
  "importRunHistory":"Importa run precedenti",
  "exportRunHistory":"Esporta run precedenti",
```

NicusPulcis commented on src/locales/it/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,36 @@
+  "viewHeldItems":"Oggetti equipaggiati ",
+  "hallofFameText":"Benvenuto alla Sala d'Onore!",
+  "viewHallOfFame":"Visualizza Sala d'Onore!",
+  "viewEndingSplash":"View ending art!"
```
```suggestion
  "viewEndingSplash":"Visualizza arte finale!"
```

NicusPulcis reviewed (approved):
italian localization done :)

frutescens commented:
Buttons have been changed!
![image](https://github.com/user-attachments/assets/836abefd-8204-44ce-aff3-d164639ceb6b)

EnochG1 commented on src/locales/ko/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,36 @@
+  "viewHeldItems":"도구",
+  "hallofFameText":"전당 등록을 축하합니다!",
+  "viewHallOfFame":"전당 보기",
+  "viewEndingSplash":"View ending art!"
```
```suggestion
  "viewEndingSplash":"엔딩 아트 보기"
```

Adri1 commented:
It is possible to have a screenshots for each language in the initial comment?

Adri1 commented:
> Hi @Adri1 I posted new photos. I will wait for new localizations, if possible, and then adjust the code so that other languages fit and look better.

Thanks a lot for it 🙏

Adri1 commented:
In the locale files, it is possible to make a two different entries for each "defeated" entry and for "hallofFameText", that depend on the player's gender?
Because in latin-based languages (and probably also German?), these entries are affected by gender (and the hallofFameText by Portugese only)

Exemple in French :
"Battu par" (male version)
"Battue par" (female version)

Not a big deal if not, but would look nicer

40chyan commented on src/locales/zh_CN/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,36 @@
+  "SPDshortened":"速率",
+  "runInfo":"游戏记录",
+  "money":"金钱",
+  "runLength":"游戏",
```
```suggestion
  "runLength":"游戏时长",
```

40chyan commented on src/locales/zh_CN/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,36 @@
+  "viewHeldItems":"持有道具",
+  "hallofFameText":"欢迎来到名人堂！",
+  "viewHallOfFame":"浏览名人堂！",
+  "viewEndingSplash":"View ending art!"
```
```suggestion
  "viewEndingSplash":"浏览结算画面"
```

40chyan reviewed (commented):
fix zh_cn

40chyan commented:
![image](https://github.com/user-attachments/assets/b14894e7-554e-4c6a-8780-579d01293944)
the position of this text is really weird
is this text even centered?

40chyan commented on src/locales/zh_CN/menu-ui-handler.ts:
```diff
@@ -16,6 +17,8 @@ export const menuUiHandler: SimpleTranslationEntries = {
   "importSlotSelect": "选择要导入到的存档位。",
   "exportSession": "导出存档",
   "exportSlotSelect": "选择要导出的存档位。",
+  "importRunHistory":"Import Run History",
+  "exportRunHistory":"Export Run History",
```
```suggestion
  "importRunHistory":"导入历史记录",
  "exportRunHistory":"导出历史记录",
```

frutescens commented:
@40chyan Hi - I will fix the off-centered text but I want to wait a day for updated localizations.

protimita commented on src/locales/ja/menu-ui-handler.ts:
```diff
@@ -16,6 +17,8 @@ export const menuUiHandler: SimpleTranslationEntries = {
   "importSlotSelect": "Select a slot to import to.",
   "exportSession": "セッションのエクスポート",
   "exportSlotSelect": "Select a slot to export from.",
+  "importRunHistory":"Import Run History",
+  "exportRunHistory":"Export Run History",
```
```suggestion
  "importRunHistory":"ラン履歴のインポート",
  "exportRunHistory":"ラン履歴のエクスポート",
```

protimita replied:
```suggestion
  "importRunHistory":"ラン履歴インポート",
  "exportRunHistory":"ラン履歴エクスポート",
```

protimita commented on src/locales/ja/menu-ui-handler.ts:
```diff
@@ -4,6 +4,7 @@ export const menuUiHandler: SimpleTranslationEntries = {
   "GAME_SETTINGS": "せってい",
   "ACHIEVEMENTS": "じっせき",
   "STATS": "とうけい",
+  "RUN_HISTORY":"Run History",
```
```suggestion
  "RUN_HISTORY":"ラン履歴",
```

protimita replied:
```suggestion
  "RUN_HISTORY":"ラン履歴",
```

protimita reviewed (commented):
JA translations for 3 lines.
menu-ui-handler.ts will be further edited on later PR

protimita commented on src/locales/ja/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,36 @@
+import { SimpleTranslationEntries } from "#app/interfaces/locales";
+
+export const runHistory: SimpleTranslationEntries = {
+  "victory": "Victory!",
+  "defeatedWild": "Defeated by ",
+  "defeatedTrainer": "Defeated by ",
+  "defeatedTrainerDouble": "Defeated by Duo",
+  "defeatedRival": "Defeated by Rival",
+  "defeated":"Defeated",
+  "luck":"Luck",
+  "score":"Score",
+  "mode":"Mode",
+  "challengeRules":"Rule(s)",
+  "challengeMonoGen1":"Gen I",
+  "challengeMonoGen2":"Gen II",
+  "challengeMonoGen3":"Gen III",
+  "challengeMonoGen4":"Gen IV",
+  "challengeMonoGen5":"Gen V",
+  "challengeMonoGen6":"Gen VI",
+  "challengeMonoGen7":"Gen VII",
+  "challengeMonoGen8":"Gen VIII",
+  "challengeMonoGen9":"Gen IX",
+  "playerItems":"Player Items",
+  "personalBest":"Personal Best!",
+  "SPDshortened":"Vel.",
+  "runInfo":"Run Info",
+  "money":"Money",
+  "runLength":"Run Length",
+  "viewHeldItems":"Held Items",
+  "hallofFameText":"Welcome to the Hall of Fame!",
+  "viewHallOfFame":"View Hall of Fame!",
+  "viewEndingSplash":"View ending art!"
```
```suggestion
  "victory": "勝利！",
  "defeatedWild": "倒された相手：",
  "defeatedTrainer": "倒された相手：",
  "defeatedTrainerDouble": "倒された相手：",
  "defeatedRival": "倒された相手：",
  "defeated":"敗北",
  "luck":"運",
  "score":"スコア",
  "mode":"モード",
  "challengeRules":"チャレンジ",
  "challengeMonoGen1":"I世代",
  "challengeMonoGen2":"II世代",
  "challengeMonoGen3":"III世代",
  "challengeMonoGen4":"IV世代",
  "challengeMonoGen5":"V世代",
  "challengeMonoGen6":"VI世代",
  "challengeMonoGen7":"VII世代",
  "challengeMonoGen8":"VIII世代",
  "challengeMonoGen9":"IX世代",
  "playerItems":"プレイヤーアイテム",
  "personalBest":"自己ベスト！",
  "SPDshortened":"速さ",//
  "runInfo":"ラン情報",
  "money":"お金",
  "runLength":"ラン最高ウェーブ",
  "viewHeldItems":"手持ちアイテム",
  "hallofFameText":"殿堂へようこそ！",
  "viewHallOfFame":"殿堂登録を見る！",
  "viewEndingSplash":"クリア後のアートを見る！"
```

mercurius-00 commented on src/locales/zh_TW/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,36 @@
+import { SimpleTranslationEntries } from "#app/interfaces/locales";
+
+export const runHistory: SimpleTranslationEntries = {
+  "victory": "Victory!",
+  "defeatedWild": "Defeated by ",
+  "defeatedTrainer": "Defeated by ",
+  "defeatedTrainerDouble": "Defeated by Duo",
+  "defeatedRival": "Defeated by Rival",
+  "defeated":"Defeated",
+  "luck":"Luck",
+  "score":"Score",
+  "mode":"Mode",
+  "challengeRules":"Rule(s)",
+  "challengeMonoGen1":"Gen I",
+  "challengeMonoGen2":"Gen II",
+  "challengeMonoGen3":"Gen III",
+  "challengeMonoGen4":"Gen IV",
+  "challengeMonoGen5":"Gen V",
+  "challengeMonoGen6":"Gen VI",
+  "challengeMonoGen7":"Gen VII",
+  "challengeMonoGen8":"Gen VIII",
+  "challengeMonoGen9":"Gen IX",
+  "playerItems":"Player Items",
+  "personalBest":"Personal Best!",
+  "SPDshortened":"Vel.",
+  "runInfo":"Run Info",
+  "money":"Money",
+  "runLength":"Run Length",
+  "viewHeldItems":"Held Items",
+  "hallofFameText":"Welcome to the Hall of Fame!",
+  "viewHallOfFame":"View Hall of Fame!",
+  "viewEndingSplash":"View ending art!"
```
```suggestion
  "victory": "勝利！",
  "defeatedWild": "被打敗",
  "defeatedTrainer": "被打敗",
  "defeatedTrainerDouble": "被組合打敗",
  "defeatedRival": "被勁敵打敗",
  "defeated":"被打敗",
  "luck":"幸運",
  "score":"分數",
  "mode":"模式",
  "challengeRules":"規則",
  "challengeMonoGen1":"I世代",
  "challengeMonoGen2":"II世代",
  "challengeMonoGen3":"III世代",
  "challengeMonoGen4":"IV世代",
  "challengeMonoGen5":"V世代",
  "challengeMonoGen6":"VI世代",
  "challengeMonoGen7":"VII世代",
  "challengeMonoGen8":"VIII世代",
  "challengeMonoGen9":"IX世代",
  "playerItems":"玩家道具",
  "personalBest":"個人最佳！",
  "SPDshortened":"速率",
  "runInfo":"遊戲記錄",
  "money":"金錢",
  "runLength":"遊戲時長",
  "viewHeldItems":"持有道具",
  "hallofFameText":"歡迎來到名人堂！",
  "viewHallOfFame":"浏覽名人堂！",
  "viewEndingSplash":"浏覽結算畫面"
```

mercurius-00 commented on src/locales/zh_TW/menu-ui-handler.ts:
```diff
@@ -4,6 +4,7 @@ export const menuUiHandler: SimpleTranslationEntries = {
   "GAME_SETTINGS": "遊戲設置",
   "ACHIEVEMENTS": "成就",
   "STATS": "數據",
+  "RUN_HISTORY": "Run History",
```
```suggestion
  "RUN_HISTORY": "歷史記錄",
```

mercurius-00 commented on src/locales/zh_TW/menu-ui-handler.ts:
```diff
@@ -16,6 +17,8 @@ export const menuUiHandler: SimpleTranslationEntries = {
   "importSlotSelect": "選擇要導入到的存檔位。",
   "exportSession": "導出存檔",
   "exportSlotSelect": "選擇要導出的存檔位。",
+  "importRunHistory":"Import Run History",
+  "exportRunHistory":"Export Run History",
```
```suggestion
  "importRunHistory":"導入歷史記錄",
  "exportRunHistory":"導出歷史記錄",
```

mercurius-00 commented on src/locales/zh_CN/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,36 @@
+  "score":"分数",
+  "mode":"模式",
+  "challengeRules":"规则",
+  "challengeMonoGen1":"一代",
+  "challengeMonoGen2":"二代",
+  "challengeMonoGen3":"三代",
+  "challengeMonoGen4":"四代",
+  "challengeMonoGen5":"五代",
+  "challengeMonoGen6":"六代",
+  "challengeMonoGen7":"七代",
+  "challengeMonoGen8":"八代",
+  "challengeMonoGen9":"九代",
```
```suggestion
  "challengeMonoGen1":"I世代",
  "challengeMonoGen2":"II世代",
  "challengeMonoGen3":"III世代",
  "challengeMonoGen4":"IV世代",
  "challengeMonoGen5":"V世代",
  "challengeMonoGen6":"VI世代",
  "challengeMonoGen7":"VII世代",
  "challengeMonoGen8":"VIII世代",
  "challengeMonoGen9":"IX世代",
```

CodeTappert commented:
Those are the two problems with the german images:
![image](https://github.com/user-attachments/assets/b9377166-d331-491c-95cf-7aa2039e5670)

The rest looks fine

(Besides the text beeing too long. But i honestly would just scale it down automatically if it gets to long)

Adri1 commented:
> Those are the two problems with the german images
> The rest looks fine
> 
> (Besides the text beeing too long. But i honestly would just scale it down automatically if it gets to long)

Yeah about the trainer category, il also raised that issue earlier
And about the hour format, I guess it's because it adjusts accortigly to the parameters of the system you play on, not on the language selected in Rogue

flx-sta commented on src/system/game-data.ts:
```diff
@@ -555,6 +573,92 @@ export class GameData {
+        return {};
+      }
+    } else {
+      const lsItemKey = `runHistoryData_${loggedInUser?.username}`;
+      const lsItem = localStorage.getItem(lsItemKey);
+      if (lsItem) {
+        const cachedResponse = lsItem;
+        if (cachedResponse) {
+          const runHistory : RunHistoryData = JSON.parse(decrypt(cachedResponse, true));
+          return runHistory;
+        }
+        return {};
+      } else {
```
I mean at this point `lsItem` is basically cacheResponse. Might as well skip the unnecessary steps:

```suggestion
      const lsItemKey = `runHistoryData_${loggedInUser?.username}`;
      const cachedData = localStorage.getItem(lsItemKey);
      if (cachedData) {
        return Promise.resolve(JSON.parse(decrypt(cachedResponse, true)) as RunHistoryData);
      } else {
```

⚠️ Be aware that `return {};` is lost in the suggestion as I don't know if that still has a place then.

----

It might be worth it to wrap `JSON.parse(decrypt)` with `try/catch` since it can break fairly easy otherwise (I know this hasn't ever been done but we thrive to better)

frutescens replied:
Yes, I decided to go for a try/catch instead

flx-sta commented on src/system/game-data.ts:
```diff
@@ -555,6 +573,92 @@ export class GameData {
     });
   }
 
+  public async getRunHistoryData(scene: BattleScene): Promise<RunHistoryData> {
```
NIT: missing method docs

AsdarDevelops commented on src/locales/es/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,36 @@
+import { SimpleTranslationEntries } from "#app/interfaces/locales";
+
+export const runHistory: SimpleTranslationEntries = {
+  "victory": "Victory!",
```
```suggestion
  "victory": "¡Victoria!",
```

flx-sta commented on src/system/game-data.ts:
```diff
@@ -555,6 +573,92 @@ export class GameData {
+    }
+  }
+
+  async saveRunHistory(scene: BattleScene, runEntry : SessionSaveData, isVictory: boolean): Promise<boolean> {
```
NIT: Missing method docs

flx-sta commented on src/system/game-data.ts:
```diff
@@ -555,6 +573,92 @@ export class GameData {
+
+  async saveRunHistory(scene: BattleScene, runEntry : SessionSaveData, isVictory: boolean): Promise<boolean> {
+
+    let runHistoryData = await this.getRunHistoryData(scene);
+    if (!runHistoryData) {
+      runHistoryData = {};
+    }
```
⚠️ This is invalid by definition!

Your method at no point associates that it might return `null` or `undefined` thus a check should not be necessary! instead the method `this.getRunHistoryData(scene)` should be made sure to not return `null` or `undefined`

flx-sta commented on src/ui/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,317 @@
+
+export type RunSelectCallback = (cursor: integer) => void;
+
+export default class RunHistoryUiHandler extends MessageUiHandler {
```
NIT: missing docs

flx-sta commented on src/ui/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,317 @@
+    this.clearruns();
+  }
+
+  eraseCursor() {
+    if (this.cursorObj) {
+      this.cursorObj.destroy();
+    }
+    this.cursorObj = null;
+  }
```
NIT: Keep naming consistent

```suggestion
  destroyCursor() {
    if (this.cursorObj) {
      this.cursorObj.destroy();
    }
    this.cursorObj = null;
  }
```

flx-sta commented on src/ui/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,317 @@
+    this.cursorObj = null;
+  }
+
+  clearruns() {
```
NIT: camel-case
```suggestion
  clearRuns() {
```

flx-sta commented on src/ui/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,317 @@
+  }
+}
+
+class RunEntryContainer extends Phaser.GameObjects.Container {
```
NIT: missing docs

flx-sta commented on src/ui/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,317 @@
+          const gameOutcomeLabel = addTextObject(this.scene, 8, 5, `${i18next.t("runHistory:defeatedTrainer")+tObj.config.nameDouble+" "+tObj.getName(0, false)}`, TextStyle.WINDOW);
+          this.add(gameOutcomeLabel);
+        } else {
+          const gameOutcomeLabel = addTextObject(this.scene, 8, 5, `${i18next.t("runHistory:defeatedTrainer")+tObj.getName(0, true)}`, TextStyle.WINDOW);
```
NIT: odd use of template string.
Either
```suggestion
          const gameOutcomeLabel = addTextObject(this.scene, 8, 5, `${i18next.t("runHistory:defeatedTrainer")}${tObj.getName(0, true)}`, TextStyle.WINDOW);
```
or
```suggestion
          const gameOutcomeLabel = addTextObject(this.scene, 8, 5, i18next.t("runHistory:defeatedTrainer")  + tObj.getName(0, true), TextStyle.WINDOW);
```

flx-sta commented on src/ui/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,317 @@
+}
+
+
+interface RunEntryContainer {
+  scene: BattleScene;
+}
```
This is some weird pattern.. shouldn't this field just be part of the class? or why is this done?

frutescens replied:
This is so RunEntryContainer has access to scene. Just Phaser things.

flx-sta commented on src/ui/run-info-ui-handler.ts:
```diff
@@ -0,0 +1,731 @@
+import { Species } from "#enums/species";
+import { PlayerGender } from "#enums/player-gender";
+
+export default class GameInfoUiHandler extends UiHandler {
```
NIT: Missing doc

flx-sta commented on src/ui/run-info-ui-handler.ts:
```diff
@@ -0,0 +1,731 @@
+  }
+
+  	async setup() {
+ 		//const page = 0;
```
NIT: can this be deleted?

flx-sta commented on src/ui/run-info-ui-handler.ts:
```diff
@@ -0,0 +1,731 @@
+  	async setup() {
+ 		//const page = 0;
+ 		this.gameStatsContainer = this.scene.add.container(1, -(this.scene.game.canvas.height / 6) + 1);
+    this.modifiersModule = await import("../modifier/modifier");
```
I didn't know that's legit to do.
<sub>The more you know</sub>

flx-sta commented on src/ui/run-info-ui-handler.ts:
```diff
@@ -0,0 +1,731 @@
+    this.setCursor(0);
+
+    this.getUi().add(this.gameStatsContainer);
+    //this.updateStats();
```
Can this be deleted?

flx-sta commented on src/ui/run-info-ui-handler.ts:
```diff
@@ -0,0 +1,731 @@
+        switch (this.runInfo.enemyParty.length) {
+        case 1:
+          // Wild - Singles
+          const enemyIconContainer = this.scene.add.container(0, 0);
+          const enemyData = this.runInfo.enemyParty[0];
+          const bossStatus = enemyData.boss;
+          enemyData.boss = false;
+          enemyData["player"] = true;
+          //addPokemonIcon() throws an error if the Pokemon used is a boss
+          const enemy = enemyData.toPokemon(this.scene);
+          const enemyIcon = this.scene.addPokemonIcon(enemy, 0, 0, 0, 0);
+          const enemyLevel = addTextObject(this.scene, 36, 26, `${i18next.t("saveSlotSelectUiHandler:lv")}${Utils.formatLargeNumber(enemy.level, 1000)}`, bossStatus ? TextStyle.PARTY_RED : TextStyle.PARTY, { fontSize: "44px", color: "#f8f8f8" });
+          enemyLevel.setShadow(0, 0, undefined);
+          enemyLevel.setStroke("#424242", 14);
+          enemyLevel.setOrigin(1, 0);
+          enemyIconContainer.add(enemyIcon);
+          enemyIconContainer.add(enemyLevel);
+          enemyContainer.add(enemyIconContainer);
+          enemyContainer.setPosition(27, 12);
+          enemy.destroy();
```
Consider moving this into it's own private method then
something like `private parse wildSingleResult()`

flx-sta commented on src/ui/run-info-ui-handler.ts:
```diff
@@ -0,0 +1,731 @@
+          break;
+        case 2:
+          //Wild - Doubles
+          this.runInfo.enemyParty.forEach((enemyData, e) => {
+            const enemyIconContainer = this.scene.add.container(0, 0);
+            const bossStatus = enemyData.boss;
+            enemyData.boss = false;
+            enemyData["player"] = true;
+            const enemy = enemyData.toPokemon(this.scene);
+            const enemyIcon = this.scene.addPokemonIcon(enemy, 0, 0, 0, 0);
+            const enemyLevel = addTextObject(this.scene, 36, 26, `${i18next.t("saveSlotSelectUiHandler:lv")}${Utils.formatLargeNumber(enemy.level, 1000)}`, bossStatus ? TextStyle.PARTY_RED : TextStyle.PARTY, { fontSize: "44px", color: "#f8f8f8" });
+            enemyLevel.setShadow(0, 0, undefined);
+            enemyLevel.setStroke("#424242", 14);
+            enemyLevel.setOrigin(1, 0);
+            enemyIconContainer.add(enemyIcon);
+            enemyIconContainer.add(enemyLevel);
+            enemyIconContainer.setPosition(e*35, 0);
+            enemyContainer.add(enemyIconContainer);
+            enemy.destroy();
+          });
+          enemyContainer.setPosition(8, 14);
```
Consider moving this into it's own private method.
e.g. `private parseWildDoubleResult()`

flx-sta commented on src/ui/run-info-ui-handler.ts:
```diff
@@ -0,0 +1,731 @@
+            const teraDetails = modifier?.getArgs();
+            const pkmnId = teraDetails[0];
+            teraPokemon[pkmnId] = teraDetails[1];
+          }
+        });
+
+        const enemyPartyContainer = this.scene.add.container(0, 0);
+        this.runInfo.enemyParty.forEach((enemyData, e) => {
+          const pokemonRowHeight = Math.floor(e/3);
+          const enemyIconContainer = this.scene.add.container(0, 0);
+          enemyIconContainer.setScale(0.6);
+          const isBoss = enemyData.boss;
+          enemyData.boss = false;
+          enemyData["player"] = true;
+          const enemy = enemyData.toPokemon(this.scene);
+          const enemyIcon = this.scene.addPokemonIcon(enemy, 0, 0, 0, 0);
+          const enemySprite1 = enemyIcon.list[0] as Phaser.GameObjects.Sprite;
+          const enemySprite2 = (enemyIcon.list.length > 1) ? enemyIcon.list[1] as Phaser.GameObjects.Sprite : undefined;
+          if (teraPokemon[enemyData.id]) {
+            const teraTint = getTypeRgb(teraPokemon[enemyData.id]);
+            const teraColor = new Phaser.Display.Color(teraTint[0], teraTint[1], teraTint[2]);
+            enemySprite1.setTint(teraColor.color);
+            if (enemySprite2) {
+              enemySprite2.setTint(teraColor.color);
+            }
+          }
+          enemyIcon.setPosition(39*(e%3)+5, (35*pokemonRowHeight));
+          const enemyLevel = addTextObject(this.scene, 43*(e%3), (27*(pokemonRowHeight+1)), `${i18next.t("saveSlotSelectUiHandler:lv")}${Utils.formatLargeNumber(enemy.level, 1000)}`, isBoss ? TextStyle.PARTY_RED : TextStyle.PARTY, { fontSize: "54px" });
+          enemyLevel.setShadow(0, 0, undefined);
+          enemyLevel.setStroke("#424242", 14);
+          enemyLevel.setOrigin(0, 0);
+
+          enemyIconContainer.add(enemyIcon);
+          enemyIconContainer.add(enemyLevel);
+          enemyPartyContainer.add(enemyIconContainer);
+          enemy.destroy();
+        });
+        enemyPartyContainer.setPosition(25, 15);
+        enemyContainer.add(enemyPartyContainer);
+      }
```
Consider moving this into it's own private method 
e.g. `private parseTrainerResult()`

flx-sta commented on src/ui/run-info-ui-handler.ts:
```diff
@@ -0,0 +1,731 @@
+    const runInfoText = addBBCodeTextObject(this.scene, 7, 0, "", TextStyle.WINDOW, {fontSize : "50px", lineSpacing:3});
+    const runTime = Utils.getPlayTimeString(this.runInfo.playTime);
+    runInfoText.appendText(`${i18next.t("runHistory:runLength")}: ${runTime}`, false);
+    runInfoText.appendText(`[color=#e8e8a8]\u20BD${Utils.formatLargeNumber(this.runInfo.money, 1000)}[/color]`);
```
NIT:
```suggestion
    runInfoText.appendText(`[color=${LIGHT_YELLOW}]\u20BD${Utils.formatLargeNumber(this.runInfo.money, 1000)}[/color]`);
```

flx-sta commented on src/ui/run-info-ui-handler.ts:
```diff
@@ -0,0 +1,731 @@
+        const icon = modifier?.getIcon(this.scene, false);
+        if (icon) {
+          const rowHeightModifier = Math.floor(visibleModifierIndex/7);
+          icon.setPosition(24 * (visibleModifierIndex%7), 20+(35*rowHeightModifier));
```
NIT: 
```suggestion
          icon.setPosition(24 * (visibleModifierIndex%7), 20 + (35 * rowHeightModifier));
```

frutescens replied:
Sorry, I absent-mindedly edited your suggestion in.

flx-sta commented on src/ui/run-info-ui-handler.ts:
```diff
@@ -0,0 +1,731 @@
+              break;
+            }
+            const itemIcon = item?.getIcon(this.scene, true);
+            //itemIcon.setFrame(item.type.iconImage);
```
Can this be removed?

flx-sta commented on src/ui/run-info-ui-handler.ts:
```diff
@@ -0,0 +1,731 @@
+  showParty(partyVisible: boolean): void {
+    const allContainers = this.partyContainer.getAll("name", "PkmnInfo");
+    allContainers.forEach((c: Phaser.GameObjects.Container) => {
+      (c.getByName("PkmnMoves") as Phaser.GameObjects.Container).setVisible(partyVisible);
+      (c.getByName("PkmnInfoText") as Phaser.GameObjects.Container).setVisible(partyVisible);
+      (c.getByName("PkmnStatsText") as Phaser.GameObjects.Container).setVisible(partyVisible);
+      (c.getByName("PkmnMarks") as Phaser.GameObjects.Container).setVisible(partyVisible);
+      (c.getByName("heldItems") as Phaser.GameObjects.Container).setVisible(!partyVisible);
```
If you look at the definition of `getByName()`:
```ts
getByName<T extends Phaser.GameObjects.GameObject>(name: string): T;
```
you can notice that passing custom interface as return type is possible through generic type definition.
You should use that

```suggestion
      c.getByName<Phaser.GameObjects.Container>("PkmnMoves").setVisible(partyVisible);
      c.getByName<Phaser.GameObjects.Container>("PkmnInfoText").setVisible(partyVisible);
      c.getByName<Phaser.GameObjects.Container>("PkmnStatsText").setVisible(partyVisible);
      c.getByName<Phaser.GameObjects.Container>("PkmnMarks").setVisible(partyVisible);
      c.getByName<Phaser.GameObjects.Container>("heldItems").setVisible(!partyVisible);
```

flx-sta commented on src/ui/run-info-ui-handler.ts:
```diff
@@ -0,0 +1,731 @@
+        pokemonSprite.setPipelineData("shiny", shiny);
+        pokemonSprite.setPipelineData("variant", variant);
+        pokemonSprite.setPipelineData("spriteKey", species.getSpriteKey(female, formIndex, shiny, variant));
+        /**
+        pkmn.resetSummonData();
+        [ "spriteColors", "fusionSpriteColors" ].map(k => {
+          delete pokemonSprite.pipelineData[`${k}Base`];
+          if (pkmn.summonData?.speciesForm) {
+            k += "Base";
+          }
+          const temp = pkmn.getSprite();
+          console.log(temp);
+          pokemonSprite.pipelineData[k] = pkmn.getSprite().pipelineData[k];
+        });*/
```
Can this be deleted?

frutescens replied:
Yes, I'll delete it.

flx-sta commented on src/ui/run-info-ui-handler.ts:
```diff
@@ -0,0 +1,731 @@
+  }
+
+
+  // The use of non-directional / A / B buttons is named in relation to functions used during starter-select.
+  // Button.CYCLE_FORM (F key) --> displays ending art (victory only)
+  // Button.CYCLE_SHINY (R key) --> displays hall of fame (victory only)
+  // Button.CYCLE_ABILITY (E key) --> shows pokemon held items
```
as @frutescens already mentioned:
Why not 
```ts
/**
 *
 */
 ```
 instead?

flx-sta reviewed (changes requested):
if you really want to bring it to the absolute next level you might wanna check if some of those new fields in the classes could be set to `readonly`. You should also check if the scopes (private/public/protected) are all correct and for all methods that are overwritten you should put a `override` in front of them to indicate that

NicusPulcis commented on src/locales/it/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,36 @@
+  "runInfo":"Info Run",
+  "money":"Patrimonio",
+  "runLength":"Durata Run",
+  "viewHeldItems":"Oggetti equipaggiati ",
+  "hallofFameText":"Benvenuto alla Sala d'Onore!",
+  "viewHallOfFame":"Visualizza Sala d'Onore!",
+  "viewEndingSplash":"Visualizza arte finale!"
```
```suggestion
  "viewHeldItems":"Oggetti equip.",
  "hallofFameText":"Benvenuto alla Sala d'Onore!",
  "viewHallOfFame":"Vai alla Sala d'Onore!",
  "viewEndingSplash":"Vai all'arte finale!"
```

NicusPulcis reviewed (commented):
![momentanea](https://github.com/user-attachments/assets/568c4912-77f0-4732-8d6e-fdf873e5ac4b)
after taking a look at this, i edited a couple of text to be better formatted and fit in the boxes : )

frutescens commented:
Fixing Incorrect Doubles titles
![image](https://github.com/user-attachments/assets/5f2e16b6-1cfc-480f-bffe-ebe3e3e60668)

frutescens commented:
Today's To-Do List
- [ ] Change text/container positioning based on text length (improving localizations)
- [ ] Break down large functions
- [ ] Write documentation
- [ ] Examine scope of variables + change to be reflective 
- [ ] Implement UI modes so the user has a more intuitive experience with exiting out of HoF/Ending Art 
- [ ] Implement more of flx-sta's suggestions

Adri1 commented on src/locales/en/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,36 @@
+  "challengeMonoGen9":"Gen IX",
+  "playerItems":"Player Items",
+  "personalBest":"Personal Best!",
+  "SPDshortened":"Vel.",
```
Question, where this actually used, or just something that is supposed to be deleted?

frutescens replied:
Are you asking about SPDshortened? It's used for Spanish and Portuguese because Veloc. is too long.

Adri1 replied:
Thanks for the clarification

frutescens commented:
@Adri1 - I apologize for the wait, but I updated the localization files to have male and female versions of 'Defeated X' and the hall of fame text. Thank you for your patience. I will eventually replace the pictures with updated versions, but the text is now dynamic and will position itself accordingly 
![image](https://github.com/user-attachments/assets/18c08655-fd10-487d-9934-157efc033167)

Adri1 commented:
Oh! You're the best thanks for that!

While checking, I noticed that it displays "Wave" in English :
![image](https://github.com/user-attachments/assets/b9387c1c-0173-4865-8700-57de1a29b1d3)
Idk if it's something forgotten of still something in progres, but I notice you about it anyway 👌

frutescens commented:
That's my bad, whoops. Lots of moving pieces. Sending a quick-fix now

Adri1 commented on src/locales/fr/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,42 @@
+  "defeatedWildM": "Battu par ",
+  "defeatedTrainerM": "Battu par ",
+  "defeatedTrainerDoubleM": "Battu par Duo",
+  "defeatedRivalM": "Battu par Rival",
+  "defeatedM":"Vaincu",
+  "defeatedWildF": "Battu par ",
+  "defeatedTrainerF": "Battu par ",
+  "defeatedTrainerDoubleF": "Battu par Duo",
+  "defeatedRivalF": "Battu par Rival",
+  "defeatedF":"Vaincu",
```
```suggestion
  "defeatedRivalM": "Battu par la Rivale",
  "defeatedM":"Vaincu",
  "defeatedWildF": "Battue par ",
  "defeatedTrainerF": "Battue par ",
  "defeatedTrainerDoubleF": "Battue par Duo",
  "defeatedRivalF": "Battue par le Rival",
  "defeatedF":"Vaincue",
```

frutescens commented:
Tomorrow's To-Do List
- [ ] Examine scope of variables + change to be reflective of their actual use
- [ ] Run Info UI Handler Documentation
- [ ] Additional Documentation
- [ ] Misc. flx-sta changes
- [ ] Check if name handling of rivals have improved 
- [ ] Testing / Screenshots
    - [ ] Upload screenshots for all languages
    - [ ] Check Wild Single Battle Defeat
    - [ ] Check Wild Double Battle Defeat
    - [ ] Check Single Trainer with < 3 Pokemon Defeat
    - [ ] Check Single Trainer with > 3 Pokemon Defeat
    - [ ] Check Double Trainer Defeat
    - [ ] Check Daily Mode Victory + Defeat
    - [ ] Check Endless Mode Victory + Defeat
    - [ ] Check Challenges - Are rules being displayed properly? 
    - [ ] Spliced Pokemon
    - [ ] Behavior handling max # of runs - should delete the oldest one

CodeTappert commented on src/locales/de/menu-ui-handler.ts:
```diff
@@ -16,6 +17,8 @@ export const menuUiHandler: SimpleTranslationEntries = {
   "importSlotSelect": "Wähle einen Slot zum Importieren.",
   "exportSession": "Sitzung exportieren",
   "exportSlotSelect": "Wähle einen Slot zum Exportieren.",
+  "importRunHistory":"Laufhistorie importieren",
+  "exportRunHistory":"Laufhistorie exportieren",
```
```suggestion
  "importRunHistory": "Laufhistorie importieren",
  "exportRunHistory": "Laufhistorie exportieren",
```

CodeTappert commented on src/locales/de/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,42 @@
+  "defeatedWildF": "Besiegt durch ",
+  "defeatedTrainerF": "Besiegt durch ",
+  "defeatedTrainerDoubleF": "Besiegt durch Doppelkampf",
+  "defeatedRivalF": "Besiegt durch Rivale",
```
```suggestion
  "defeatedRivalF": "Besiegt durch Rivalin",
```

NicusPulcis commented on src/locales/it/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,42 @@
+
+export const runHistory: SimpleTranslationEntries = {
+  "victory": "Vittoria!",
+  "defeatedWildM": "Sconfitti da ",
+  "defeatedTrainerM": "Sconfitti da ",
+  "defeatedTrainerDoubleM": "Sconfitti dalla coppia ",
+  "defeatedRivalM": "Sconfitti dal rivale",
+  "defeatedM":"Sconfitti",
+  "defeatedWildF": "Sconfitti da ",
+  "defeatedTrainerF": "Sconfitti da ",
+  "defeatedTrainerDoubleF": "Sconfitti dalla coppia ",
+  "defeatedRivalF": "Sconfitti dal rivale",
+  "defeatedF":"Sconfitti",
```
```suggestion
  "defeatedWildM": "Sconfitto da ",
  "defeatedTrainerM": "Sconfitto da ",
  "defeatedTrainerDoubleM": "Sconfitto dalla coppia ",
  "defeatedRivalM": "Sconfitto dal rivale",
  "defeatedM":"Sconfitto",
  "defeatedWildF": "Sconfitta da ",
  "defeatedTrainerF": "Sconfitta da ",
  "defeatedTrainerDoubleF": "Sconfitta dalla coppia ",
  "defeatedRivalF": "Sconfitta dal rivale",
  "defeatedF":"Sconfitta",
```

NicusPulcis reviewed (approved):
italian male/female thingy fixed :) thanks for the new codes @frutescens

CodeTappert commented on src/locales/de/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,42 @@
+  "defeatedWildM": "Besiegt durch ",
+  "defeatedTrainerM": "Besiegt durch ",
+  "defeatedTrainerDoubleM": "Besiegt durch Doppelkampf",
+  "defeatedRivalM": "Besiegt durch Rivale",
```
```suggestion
  "defeatedRivalM": "Besiegt durch Rivalin",
```

CodeTappert replied:
@frutescens Since those keys refer to the player gender apparently...
Make it more clear. For dialogue for example we use PGM and PGF (playerGenderMale and playerGenderFemale) 
so maybe "defeatedRivalPGM" etc

flx-sta replied:
FYI a cleaner solution would be to use the `context` i18n object
https://www.i18next.com/translation-function/context#basic

```json
{
      "friend": "A friend",
      "friend_male": "A boyfriend",
      "friend_female": "A girlfriend"
}
```
```ts
i18next.t('friend'); // -> "A friend"
i18next.t('friend', { context: 'male' }); // -> "A boyfriend"
i18next.t('friend', { context: 'female' }); // -> "A girlfriend"
```

AsdarDevelops commented on src/locales/es/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,42 @@
+  "runLength":"Duración de partida",
+  "viewHeldItems":"Objetos equipados",
+  "hallofFameTextM":"¡Bienvenido al Hall de la Fama!",
+  "hallofFameTextF":"¡Bienvenido al Hall de la Fama!",
```
```suggestion
  "hallofFameTextF":"¡Bienvenida al Hall de la Fama!",
```

AsdarDevelops commented on src/locales/es/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,42 @@
+  "defeatedTrainerDoubleM": "Derrotado por un dúo",
+  "defeatedRivalM": "Derrotado por el rival",
+  "defeatedM":"Derrotado",
+  "defeatedWildF": "Derrotado por ",
+  "defeatedTrainerF": "Derrotado por ",
+  "defeatedTrainerDoubleF": "Derrotado por un dúo",
```
```suggestion
  "defeatedWildF": "Derrotada por ",
  "defeatedTrainerF": "Derrotada por ",
  "defeatedTrainerDoubleF": "Derrotada por un dúo",
```

AsdarDevelops commented on src/locales/es/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,42 @@
+  "defeatedWildF": "Derrotado por ",
+  "defeatedTrainerF": "Derrotado por ",
+  "defeatedTrainerDoubleF": "Derrotado por un dúo",
+  "defeatedRivalF": "Derrotado por el rival",
+  "defeatedF":"Derrotado",
```
```suggestion
  "defeatedRivalF": "Derrotada por el rival",
  "defeatedF":"Derrotada",
```

AsdarDevelops reviewed (commented):
Added female protag texts

josericardo-fo commented on src/locales/pt_BR/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,42 @@
+  "defeatedTrainerDoubleM": "Derrotado por Dupla",
+  "defeatedRivalM": "Derrotado por Rival",
+  "defeatedM": "Derrotado",
+  "defeatedWildF": "Derrotado por ",
+  "defeatedTrainerF": "Derrotado por ",
+  "defeatedTrainerDoubleF": "Derrotado por Dupla",
+  "defeatedRivalF": "Derrotado por Rival",
+  "defeatedF": "Derrotado",
```
```suggestion
  "defeatedWildF": "Derrotada por ",
  "defeatedTrainerF": "Derrotada por ",
  "defeatedTrainerDoubleF": "Derrotada por Dupla",
  "defeatedRivalF": "Derrotada por Rival",
  "defeatedF": "Derrotada",
```

Adri1 commented on src/locales/fr/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,42 @@
+  "defeatedWildM": "Battu par ",
+  "defeatedTrainerM": "Battu par ",
+  "defeatedTrainerDoubleM": "Battu par Duo",
+  "defeatedRivalM": "Battu par la Rivale",
+  "defeatedM":"Vaincu",
+  "defeatedWildF": "Battue par ",
+  "defeatedTrainerF": "Battue par ",
+  "defeatedTrainerDoubleF": "Battue par Duo",
+  "defeatedRivalF": "Battue par le Rival",
```
```suggestion
  "defeatedRivalM": "Battu par Rivale",
  "defeatedM":"Vaincu",
  "defeatedWildF": "Battue par ",
  "defeatedTrainerF": "Battue par ",
  "defeatedTrainerDoubleF": "Battue par Duo",
  "defeatedRivalF": "Battue par Rival",
```

Adri1 reviewed (commented):
Tiny French ajustements after seeing ```defeatedRival``` entries in a context

Adri1 commented:
Would it be possible when you can, to have the run histroy menu with an entry where you are defeted by a wild Pokémon please ?🙏

frutescens commented:
@Adri1 
![image](https://github.com/user-attachments/assets/be51d0ae-8e3b-4190-a8d8-80aa62a4bce3)

Adri1 commented:
@frutescens I was more thinking about this screen :
![358781811-1ad32ad1-e722-4abf-a3fc-2e4986c23bf8](https://github.com/user-attachments/assets/a5643098-f94d-4c1e-8a87-e8d6d3d65cfb)

frutescens commented:
Ah, my bad @Adri1 
![image](https://github.com/user-attachments/assets/03c4b87b-c06a-476a-8221-d66f01b2b417)

frutescens commented:
Some of the Fusion names for non-English languages are very long. I believe that this PR should fix some of the issues. https://github.com/pagefaultgames/pokerogue/pull/3262

Vassiat commented on src/ui/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,361 @@
+      this.scene.add.existing(entry);
+      this.runsContainer.add(entry);
+      this.runs.push(entry);
+    }
```
Cursor is rendered before this for loop, so it is below windows

Vassiat reviewed (changes requested):
Cursor is rendered before this for loop, so it is below windows

Vassiat commented on src/ui/run-history-ui-handler.ts:
```diff
@@ -0,0 +1,361 @@
+      this.scene.add.existing(entry);
+      this.runsContainer.add(entry);
+      this.runs.push(entry);
+    }
```
```suggestion
    }
    if (this.cursorObj) {
      this.runsContainer.bringToTop(this.cursorObj);
      }
```

frutescens replied:
wait... omg thank you so much

Vassiat commented:
I think it should be changed from where it is accessed, from `menu-ui-handler` to `title-phase`, and with it some locale from menuUiHandler to menu, because the log out is not being considered.

<img src="https://github.com/user-attachments/assets/bad248ef-f57c-47f6-92a6-2ede7ae7cd03"
         alt="image" />

<img src="https://github.com/user-attachments/assets/8104037c-fe8c-4791-b999-a6b6112cd1f3"
         alt="image" />

frutescens commented:
ah, I have a PR that will combine both vouchers and achievements together so there will be room

Vassiat commented on src/ui/menu-ui-handler.ts:
Due to the incorporation of importing/exporting runHistory, the MANAGE_DATA modal is a bit out of place, making it impossible to view the import/export session: 

![image](https://github.com/user-attachments/assets/9421077f-2bdb-4501-a7a2-6be9cbf80623)
> ### _This would only be seen in beta, as in production the import option is removed and in development the links_

### For this, I found two solutions:
- Add maxOptions to that modal:
``` ts
this.manageDataConfig = {
      xOffset: 98,
      options: manageDataOptions,
      maxOptions: 6
};
```
![image](https://github.com/user-attachments/assets/86697618-b3a6-48bc-baff-6aa81df9d3a5)

- Or make it scroll down depending on the number of options:
``` ts
//  if left is undefined, right assign
this.manageDataConfig.yOffset ??= this.manageDataConfig.options.length ** 1.66;

ui.setOverlayMode(Mode.MENU_OPTION_SELECT, this.manageDataConfig);
success = true;
```
![image](https://github.com/user-attachments/assets/11e095de-9bf5-4eaa-85a0-52fee139beb4)

frutescens commented:
![image](https://github.com/user-attachments/assets/defe6d91-cd97-43f6-b33f-3bb4c6ad5c14)

frutescens commented:
![image](https://github.com/user-attachments/assets/fbc2a29e-6940-4fff-b1e1-9a2ad1c5d18a)
